# Mục lục

##  1.Import library
##  2.Data Processing
- [2.1. Import dữ liệu](#21-import-dữ-liệu)
  - [2.1.1. HUYỆN CƯ JUT](#211-huyện-cư-jut)
  - [2.1.2. HUYỆN DAK GLONG](#212-huyện-dak-glong)
  - [2.1.3. HUYỆN DAK MIL](#213-huyện-dak-mil)
  - [2.1.4. HUYỆN DAK RLAP](#214-huyện-dak-rlap)
  - [2.1.5. HUYỆN DAK SONG](#215-huyện-dak-song)
  - [2.1.6. HUYỆN KRONG NO](#216-huyện-krong-no)
  - [2.1.7. HUYỆN TUY ĐỨC](#217-huyện-tuy-đức)
  - [2.1.8. TP GIA NGHĨA](#218-tp-gia-nghĩa)
- [2.2. Tiền xử lý dữ liệu](#22-tiền-xử-lý-dữ-liệu)
  - [2.2.1. HUYỆN CƯ JUT](#221-huyện-cư-jut)
  - [2.2.2. HUYỆN DAK GLONG](#222-huyện-dak-glong)
  - [2.2.3. HUYỆN DAK MIL](#223-huyện-dak-mil)
  - [2.2.4. HUYỆN DAK RLAP](#224-huyện-dak-rlap)
  - [2.2.5. HUYỆN DAK SONG](#225-huyện-dak-song)
  - [2.2.6. HUYỆN KRONG NO](#226-huyện-krong-no)
  - [2.2.7. HUYỆN TUY ĐỨC](#227-huyện-tuy-đức)
  - [2.2.8. TP GIA NGHĨA](#228-tp-gia-nghĩa)


## 1.Import library

In [245]:
import pandas as pd
pd.set_option('display.max_rows', None)
from datetime import datetime
pd.set_option('display.max_columns', None)

## 2.Data Processing


### 2.1 Import DataSet

####  2.1.1 HUYỆN CU JUT

In [246]:
df_huyen_cu_jut=pd.read_excel('../data_origin/HUYỆN CƯ JUT.xlsx')

> năm dòng đầu tiên

In [247]:
df_huyen_cu_jut.head(5)

Unnamed: 0  STT_HO  STT     HO_TEN_CHU_HO  HO_TEN_THANH_VIEN  \
0           1     1.0    1    TRƯƠNG THỊ LÙN     TRƯƠNG THỊ LÙN   
1           2     2.0    2  HOÀNG VĂN TRƯỜNG   HOÀNG VĂN TRƯỜNG   
2           3     NaN    3  HOÀNG VĂN TRƯỜNG       PHẠM THỊ SOI   
3           4     NaN    4  HOÀNG VĂN TRƯỜNG  HOÀNG THỊ BẢO THI   
4           5     NaN    5  HOÀNG VĂN TRƯỜNG       HOÀNG KỲ ANH   

   QUANHE\n_CHUHO   NGAY_SINH NGAY_SINH.1  Unnamed: 8 Unnamed: 9  Unnamed: 10  \
0             1.0  22/04/1905          22         4.0          /       1905.0   
1             1.0  02/06/1905          02         6.0          /       1905.0   
2             2.0  03/06/1905          03         6.0          /       1905.0   
3             3.0  03/07/1905          03         7.0          /       1905.0   
4             3.0  05/07/1905          05         7.0          /       1905.0   

   Unnamed: 11  GIOI_TINH       SO_CCCD      TINH   HUYEN         XA  \
0          NaN        2.0  054139002355  ĐẮK NÔNG  CƯ JUT  TÂM THẮNG   
1          NaN        1.0  036080016941  ĐẮK NÔNG  CƯ JUT  TÂM THẮNG   
2          NaN        2.0  036181021986  ĐẮK NÔNG  CƯ JUT  TÂM THẮNG   
3          NaN        2.0  067311001529  ĐẮK NÔNG  CƯ JUT  TÂM THẮNG   
4          NaN        1.0  067213007898  ĐẮK NÔNG  CƯ JUT  TÂM THẮNG   

   THÔN TỔ ẤP DAN \nTOC  PHAN_\nLOAI_HO  \
0         NaN         1               1   
1         NaN         1               1   
2         NaN         1               1   
3         NaN         1               1   
4         NaN         1               1   

                                          KET_QUA GHI_CHU   DC_DINHDANH  \
0                              TIM THAY THEO CCCD     NaN  054139002355   
1                              TIM THAY THEO CCCD     NaN  036080016941   
2                              TIM THAY THEO CMND     NaN  036181021986   
3  TIM THAY THEO HO TEN, NAMSINH, TINH, HUYEN, XA     NaN  067311001529   
4  TIM THAY THEO HO TEN, NAMSINH, TINH, HUYEN, XA     NaN  067213007898   

     DC_CMND  Unnamed: 24 Unnamed: 25  
0  245119301          NaN         NaN  
1  245119293          NaN         NaN  
2  245434228          NaN         NaN  
3        NaN          NaN         NaN  
4        NaN          NaN         NaN

In [248]:
df_huyen_cu_jut.columns

Index(['Unnamed: 0', 'STT_HO', 'STT', 'HO_TEN_CHU_HO', 'HO_TEN_THANH_VIEN',
       'QUANHE\n_CHUHO', 'NGAY_SINH', 'NGAY_SINH.1', 'Unnamed: 8',
       'Unnamed: 9', 'Unnamed: 10', 'Unnamed: 11', 'GIOI_TINH', 'SO_CCCD',
       'TINH', 'HUYEN', 'XA', 'THÔN TỔ ẤP', 'DAN \nTOC', 'PHAN_\nLOAI_HO',
       'KET_QUA', 'GHI_CHU', 'DC_DINHDANH', 'DC_CMND', 'Unnamed: 24',
       'Unnamed: 25'],
      dtype='object')

In [249]:
df_huyen_cu_jut.shape

(11841, 26)

- ##### Bộ dữ liệu về huyện cư jut có tổng cộng 11 841 dòng và 26 thuộc tính

####  2.1.2 HUYỆN DAK GLONG

In [250]:
df_huyen_dak_glong=pd.read_excel("../data_origin/Huyện Đắk Glong.xlsx")

> 5 dòng đầu tiên

In [251]:
df_huyen_dak_glong.head(5)

Unnamed: 0 STT_HO  STT HO_TEN_CHU_HO HO_TEN_THANH_VIEN  QUANHE_CHUHO  \
0           1      1    1   H' MĂNG (B)       H' MĂNG (B)           1.0   
1           2    NaN    2   H' MĂNG (B)            H' VÂN           3.0   
2           3    NaN    3   H' MĂNG (B)           K' LINH           3.0   
3           4    NaN    4   H' MĂNG (B)           H' LUẪN           3.0   
4           5    NaN    5   H' MĂNG (B)    ĐỖ K' TUẤN ANH           5.0   

    NGAY_SINH  Unnamed: 7  Unnamed: 8  Unnamed: 9  Unnamed: 10  Unnamed: 11  \
0  1962-10-09         NaN         NaN         NaN          NaN          NaN   
1  18/03/1993         NaN         NaN         NaN          NaN          NaN   
2  15/07/1995         NaN         NaN         NaN          NaN          NaN   
3  1998-10-04         NaN         NaN         NaN          NaN          NaN   
4  2014-05-03         NaN         NaN         NaN          NaN          NaN   

   GIOI_TINH       SO_CCCD      TINH      HUYEN        XA  DANTOC  \
0        2.0  067162002471  ĐẮK NÔNG  ĐẮK GLONG  ĐẮK PLAO    28.0   
1        2.0  067193004039  ĐẮK NÔNG  ĐẮK GLONG  ĐẮK PLAO    28.0   
2        1.0  067095003896  ĐẮK NÔNG  ĐẮK GLONG  ĐẮK PLAO    28.0   
3        2.0  067198003486  ĐẮK NÔNG  ĐẮK GLONG  ĐẮK PLAO    28.0   
4        1.0  067214003650  ĐẮK NÔNG  ĐẮK GLONG  ĐẮK PLAO    28.0   

   PHAN_LOAI_HO             KET_QUA   DC_DINHDANH      DC_CMND  \
0             1  TIM THAY THEO CCCD  6.716200e+10  245076818.0   
1             1  TIM THAY THEO CCCD  6.719300e+10  245286801.0   
2             1  TIM THAY THEO CCCD  6.709500e+10  245316352.0   
3             1  TIM THAY THEO CCCD  6.719800e+10  245317786.0   
4             1  TIM THAY THEO CCCD  6.721400e+10          NaN   

                            GHI_CHU  
0  LECH THONG TIN HO TEN, NGAY SINH  
1                               NaN  
2                               NaN  
3                               NaN  
4                               NaN

In [252]:
df_huyen_dak_glong.columns

Index(['Unnamed: 0', 'STT_HO', 'STT', 'HO_TEN_CHU_HO', 'HO_TEN_THANH_VIEN',
       'QUANHE_CHUHO', 'NGAY_SINH', 'Unnamed: 7', 'Unnamed: 8', 'Unnamed: 9',
       'Unnamed: 10', 'Unnamed: 11', 'GIOI_TINH', 'SO_CCCD', 'TINH', 'HUYEN',
       'XA', 'DANTOC', 'PHAN_LOAI_HO', 'KET_QUA', 'DC_DINHDANH', 'DC_CMND',
       'GHI_CHU'],
      dtype='object')

In [253]:
df_huyen_dak_glong.shape

(37427, 23)

- ##### Bộ dữ liệu về huyện Dắk Glong có tổng cộng 37 427 và 23 thuộc tính


#### 2.1.3 HUYỆN DAK MIL 

In [254]:
df_huyen_dak_mil = pd.read_excel("../data_origin/HUYỆN DAK MIL.xlsx",sheet_name='Tổng hợp')

> 5 dòng đầu tiên

In [255]:
df_huyen_dak_mil.head(5)

NAM RA SOAT  STT_HO  STT LOẠI HỘ       Khu vực      MÃ THÔN           THÔN  \
0         2022     1.0    1      NEW  COUNTRYSIDE  tdp1-dakmil  TỔ DÂN PHỐ 01   
1         2022     NaN    2      NEW  COUNTRYSIDE  tdp1-dakmil  TỔ DÂN PHỐ 01   
2         2022     1.0    3      NEW  COUNTRYSIDE  tdp3-dakmil  TỔ DÂN PHỐ 03   
3         2022     NaN    4      NEW  COUNTRYSIDE  tdp3-dakmil  TỔ DÂN PHỐ 03   
4         2022     NaN    5      NEW  COUNTRYSIDE  tdp3-dakmil  TỔ DÂN PHỐ 03   

     HO_TEN_CHU_HO HO_TEN_THANH_VIEN      SO_CCCD QUANHE_CHUHO  NGAY_SINH  \
0      LÊ THỊ LINH       LÊ THỊ LINH  40173037183         HEAD     1973.0   
1      LÊ THỊ LINH  ĐOÀN THỊ ANH THƯ  67309000070        CHILD     2009.0   
2  NGUYỄN THỊ KHOA   NGUYỄN THỊ KHOA  67159003545         HEAD     1959.0   
3  NGUYỄN THỊ KHOA  ĐỖ THỊ MINH LIÊN  67188003929        CHILD     1988.0   
4  NGUYỄN THỊ KHOA   ĐỖ THỊ MINH NHI  67191002534        OTHER     2005.0   

  Tháng  Ngày DANTOC GIOI_TINH      TINH PHAN_LOAI_HO    HUYEN  \
0    10   9.0    KIN    FEMALE  ĐẮK NÔNG         POOR  Đắk Mil   
1    10  10.0    KIN    FEMALE  ĐẮK NÔNG         POOR  Đắk Mil   
2     1   1.0    KIN    FEMALE  ĐẮK NÔNG         POOR  Đắk Mil   
3     6  22.0    KIN    FEMALE  ĐẮK NÔNG         POOR  Đắk Mil   
4     9  20.0    KIN    FEMALE  ĐẮK NÔNG         POOR  Đắk Mil   

                 XA  
0  Thị trấn Đắk Mil  
1  Thị trấn Đắk Mil  
2  Thị trấn Đắk Mil  
3  Thị trấn Đắk Mil  
4  Thị trấn Đắk Mil

In [256]:
df_huyen_dak_mil.columns

Index(['NAM RA SOAT', 'STT_HO', 'STT', 'LOẠI HỘ ', 'Khu vực', 'MÃ THÔN',
       'THÔN', 'HO_TEN_CHU_HO', 'HO_TEN_THANH_VIEN', 'SO_CCCD', 'QUANHE_CHUHO',
       'NGAY_SINH', 'Tháng', 'Ngày', 'DANTOC', 'GIOI_TINH', 'TINH',
       'PHAN_LOAI_HO', 'HUYEN', 'XA'],
      dtype='object')

In [257]:
df_huyen_dak_mil.shape

(4723, 20)

- ##### Bộ dữ liệu về huyện Dắk MIL có tổng cộng 4724 dòng và 20 thuộc tính

#### 2.1.4. HUYỆN DAK RLAP

In [258]:
df_huyen_dak_rlap=pd.read_excel("../data_origin/HUYỆN ĐẮK RLẤP.xlsx")

> 5 dòng đầu tiên

In [259]:
df_huyen_dak_rlap.head(5)

STT_HO  STT  HO_TEN_CHU_HO   HO_TEN_THANH_VIEN  QUANHE_CHUHO   NGAY_SINH  \
0      1    1    ĐÀO VĂN TÂN         ĐÀO VĂN TÂN           1.0   15/3/1980   
1    NaN    2    ĐÀO VĂN TÂN       NGUYỄN THỊ HỆ           2.0   18/8/1987   
2    NaN    3    ĐÀO VĂN TÂN  ĐÀO THỊ VIỆT TRINH           3.0   26/2/2011   
3    NaN    4    ĐÀO VĂN TÂN         ĐÀO VIỆT MỸ           3.0   18/2/2015   
4      2    5  TRẦN THỊ CHÓT       TRẦN THỊ CHÓT           1.0  11/12/1949   

   GIOI_TINH       SO_CCCD      TINH     HUYEN       XA  DANTOC  PHAN_LOAI_HO  \
0        1.0  070080001491  ĐẮK NÔNG  ĐẮK RLẤP  NHÂN CƠ     1.0             1   
1        2.0  037187013794  ĐẮK NÔNG  ĐẮK RLẤP  NHÂN CƠ     1.0             1   
2        2.0  067311002433  ĐẮK NÔNG  ĐẮK RLẤP  NHÂN CƠ     1.0             1   
3        1.0  067215002637  ĐẮK NÔNG  ĐẮK RLẤP  NHÂN CƠ     1.0             1   
4        2.0  046149007767  ĐẮK NÔNG  ĐẮK RLẤP  NHÂN CƠ     1.0             1   

                                             KET_QUA  \
0                                 TIM THAY THEO CCCD   
1                                 TIM THAY THEO CCCD   
2                                 TIM THAY THEO CCCD   
3  KHONG TIM THAY THEO HO TEN, NAMSINH, TINH, HUY...   
4                           KHONG TIM THAY THEO CCCD   

                            GHI_CHU   DC_DINHDANH    DC_CMND         GHI CHÚ   
0                               NaN  7.008000e+10  245107431              NaN  
1  LECH THONG TIN HO TEN, NGAY SINH  3.708701e+10  164267618  Điều chỉnh CCCD  
2  LECH THONG TIN HO TEN, NGAY SINH  6.721500e+10        NaN  Điều chỉnh CCCD  
3                               NaN           NaN        NaN  Điều chỉnh CCCD  
4                               NaN           NaN        NaN  Điều chỉnh CCCD

In [260]:
df_huyen_dak_rlap.columns

Index(['STT_HO', 'STT', 'HO_TEN_CHU_HO', 'HO_TEN_THANH_VIEN', 'QUANHE_CHUHO',
       'NGAY_SINH', 'GIOI_TINH', 'SO_CCCD', 'TINH', 'HUYEN', 'XA', 'DANTOC',
       'PHAN_LOAI_HO', 'KET_QUA', 'GHI_CHU', 'DC_DINHDANH', 'DC_CMND',
       'GHI CHÚ '],
      dtype='object')

In [261]:
df_huyen_dak_rlap.shape

(4241, 18)

- ##### Bộ dữ liệu về huyện HUYỆN ĐẮK RLẤP có tổng cộng 4 241 dòng và 18 thuộc tính


#### 2.1.5. HUYỆN DAK SONG

In [262]:
df_huyen_dak_song=pd.read_excel("../data_origin/HUYỆN ĐẮK SONG.xlsx")

> 5 dòng đầu tiên

In [263]:
df_huyen_dak_song.head(5)

Unnamed: 0 STT_HO  STT      HO_TEN_CHU_HO  HO_TEN_THANH_VIEN QUANHE_CHUHO  \
0          1         1.0        LÊ THỊ TƯƠI        LÊ THỊ TƯƠI         HEAD   
1          2    NaN  2.0        LÊ VĂN TOÀN        LÊ VĂN TOÀN        CHILD   
2          3    NaN  3.0   TRẦN QUYẾT THẮNG   TRẦN QUYẾT THẮNG         HEAD   
3          4    NaN  4.0     NGUYỄN THỊ THƯ     NGUYỄN THỊ THƯ      PARTNER   
4          5    NaN  5.0  TRẦN NGUYÊN THÔNG  TRẦN NGUYÊN THÔNG        CHILD   

             NGAY_SINH NGAY_SINH.1  Ngay  Tháng     Năm GIOI_TINH  \
0           1972-08-12  12/08/1972  12.0    8.0  1972.0    FEMALE   
1           1995-06-07  07/06/1995   7.0    6.0  1995.0      MALE   
2           17/12/1962  17/12/1962  17.0   12.0  1962.0      MALE   
3  1962-04-26 00:00:00  26/04/1962  26.0    4.0  1962.0    FEMALE   
4           17/12/1999  17/12/1999  17.0   12.0  1999.0      MALE   

        SO_CCCD      TINH     HUYEN       XA DANTOC PHAN_LOAI_HO  \
0  036172015492  ĐẮK NÔNG  ĐẮK SONG  ĐẮK HÒA    KIN         POOR   
1  067095005884  ĐẮK NÔNG  ĐẮK SONG  ĐẮK HÒA    KIN         POOR   
2  040062008552  ĐẮK NÔNG  ĐẮK SONG  ĐẮK HÒA    KIN         POOR   
3  046162012878  ĐẮK NÔNG  ĐẮK SONG  ĐẮK HÒA    KIN         POOR   
4  067099001783  ĐẮK NÔNG  ĐẮK SONG  ĐẮK HÒA    KIN         POOR   

              KET_QUA GHI_CHU   DC_DINHDANH      DC_CMND  
0  TIM THAY THEO CCCD     NaN  036172015492  245128541.0  
1  TIM THAY THEO CCCD     NaN  067095005884  245291946.0  
2  TIM THAY THEO CCCD     NaN  040062008552  245023847.0  
3  TIM THAY THEO CCCD     NaN  046162012878  245023848.0  
4  TIM THAY THEO CCCD     NaN  067099001783  245379140.0

In [264]:
df_huyen_dak_song.columns

Index(['Unnamed: 0', 'STT_HO', 'STT', 'HO_TEN_CHU_HO', 'HO_TEN_THANH_VIEN',
       'QUANHE_CHUHO', 'NGAY_SINH', 'NGAY_SINH.1', 'Ngay', 'Tháng', 'Năm',
       'GIOI_TINH', 'SO_CCCD', 'TINH', 'HUYEN', 'XA', 'DANTOC', 'PHAN_LOAI_HO',
       'KET_QUA', 'GHI_CHU', 'DC_DINHDANH', 'DC_CMND'],
      dtype='object')

In [265]:
df_huyen_dak_song.shape

(9028, 22)

- ##### Bộ dữ liệu về huyện Dắk SONG có tổng cộng 9 028 dòng và 22 thuộc tính


#### 2.1.6. HUYỆN KRONG NO

In [266]:
df_huyen_krong_no=pd.read_excel("../data_origin/HUYỆN KRÔNG NÔ.xlsx",sheet_name='Tổng hợp')

> 5 dòng đầu tiên

In [267]:
df_huyen_krong_no.head(5)

b STT_HO  STT HO_TEN_CHU_HO HO_TEN_THANH_VIEN QUANHE_CHUHO   NGAY_SINH  \
0  1      1    1   ĐỖ TẤN VIỆT       ĐỖ TẤN VIỆT         HEAD  14/04/1966   
1  2    NaN    2   ĐỖ TẤN VIỆT   ĐINH THỊ PHƯƠNG      PARTNER  10/05/1976   
2  3    NaN    3   ĐỖ TẤN VIỆT     ĐỖ QUỐC TRUNG        CHILD  28/10/1999   
3  4    NaN    4   ĐỖ TẤN VIỆT  ĐỖ THỊ MINH HIẾU        CHILD  13/10/2000   
4  5    NaN    5   ĐỖ TẤN VIỆT    ĐỖ THỊ MỸ TRÂM        CHILD  22/12/2002   

   Ngay  tháng   năm GIOI_TINH       SO_CCCD      TINH     HUYEN       XA  \
0  14.0    4.0  1966      MALE  049066010173  ĐẮK NÔNG  KRÔNG NÔ  ĐẮK MÂM   
1  10.0    5.0  1976    FEMALE  042176007680  ĐẮK NÔNG  KRÔNG NÔ  ĐẮK MÂM   
2  28.0   10.0  1999      MALE  067099006144  ĐẮK NÔNG  KRÔNG NÔ  ĐẮK MÂM   
3  13.0   10.0  2000    FEMALE  067300000133  ĐẮK NÔNG  KRÔNG NÔ  ĐẮK MÂM   
4  22.0   12.0  2002    FEMALE  067302005444  ĐẮK NÔNG  KRÔNG NÔ  ĐẮK MÂM   

  DANTOC PHAN_LOAI_HO             KET_QUA                           GHI_CHU  \
0    KIN         POOR  TIM THAY THEO CCCD                               NaN   
1    KIN         POOR  TIM THAY THEO CCCD                               NaN   
2    KIN         POOR  TIM THAY THEO CCCD                               NaN   
3    KIN         POOR  TIM THAY THEO CCCD                               NaN   
4    KIN         POOR  TIM THAY THEO CCCD  LECH THONG TIN HO TEN, NGAY SINH   

    DC_DINHDANH    DC_CMND  
0  049066010173  245090459  
1  042176007680  245090460  
2  067099006144  245356836  
3  067300000133  245356837  
4  067302005444  245419754

In [268]:
df_huyen_krong_no.columns

Index(['b', 'STT_HO', 'STT', 'HO_TEN_CHU_HO', 'HO_TEN_THANH_VIEN',
       'QUANHE_CHUHO', 'NGAY_SINH', 'Ngay', 'tháng', 'năm', 'GIOI_TINH',
       'SO_CCCD', 'TINH', 'HUYEN', 'XA', 'DANTOC', 'PHAN_LOAI_HO', 'KET_QUA',
       'GHI_CHU', 'DC_DINHDANH', 'DC_CMND'],
      dtype='object')

In [269]:
df_huyen_krong_no.shape

(15094, 21)

- ##### Bộ dữ liệu về huyện KRONG NO có tổng cộng 15 094 dòng và 21 thuộc tính


#### 2.1.7. HUYỆN TUY ĐỨC 

In [270]:
df_huyen_tuy_duc=pd.read_excel("../data_origin/HUYỆN TUY ĐỨC.xlsx")

> 5 dòng đầu tiên

In [271]:
df_huyen_tuy_duc.head(5)

Unnamed: 0 STT_HO  STT    HO_TEN_CHU_HO HO_TEN_THANH_VIEN QUANHE_CHUHO  \
0           1      1    1      HÀ VĂN CÔNG       HÀ VĂN CÔNG         HEAD   
1           2    NaN    2      HÀ VĂN CÔNG      NGÔ THỊ NƯỚC      PARTNER   
2           3    NaN    3      HÀ VĂN CÔNG        HÀ VĂN HỮU        CHILD   
3           4      2    4  PHẠM TRUNG HIẾU   PHẠM TRUNG HIẾU         HEAD   
4           5    NaN    5  PHẠM TRUNG HIẾU      PHẠM THỊ HẬU       PARENT   

    NGAY_SINH NGAY_SINH.1  NGAY  Tháng   Năm GIOI_TINH       SO_CCCD  \
0  20/05/1952  20/05/1952  20.0    5.0  1952    FEMALE  034052015155   
1  20/10/1952  20/10/1952  20.0   10.0  1952    FEMALE     245069748   
2  24/02/1990  24/02/1990  24.0    2.0  1990      MALE  chưa có CCCD   
3  20/06/1998  20/06/1998  20.0    6.0  1998      MALE  067098005941   
4        1970        1970   NaN    NaN  1970    FEMALE     245323659   

       TINH    HUYEN         XA  THÔN XÓM DANTOC PHAN_LOAI_HO  \
0  ĐẮK NÔNG  TUY ĐỨC  QUẢNG TÂN       NaN    KIN         POOR   
1  ĐẮK NÔNG  TUY ĐỨC  QUẢNG TÂN       NaN    KIN         POOR   
2  ĐẮK NÔNG  TUY ĐỨC  QUẢNG TÂN       NaN    KIN         POOR   
3  ĐẮK NÔNG  TUY ĐỨC  QUẢNG TÂN       NaN    KIN         POOR   
4  ĐẮK NÔNG  TUY ĐỨC  QUẢNG TÂN       NaN    KIN         POOR   

                    KET_QUA GHI_CHU   DC_DINHDANH      DC_CMND Unnamed: 23  
0        TIM THAY THEO CCCD     NaN  034052015155  245069749.0           1  
1        TIM THAY THEO CMND     NaN  034152019073  245069748.0           2  
2  KHONG TIM THAY THEO CCCD     NaN           NaN          NaN           3  
3        TIM THAY THEO CCCD     NaN  067098005941  245377043.0           4  
4        TIM THAY THEO CMND     NaN  034170017399  245323659.0           5

In [272]:
df_huyen_tuy_duc.columns

Index(['Unnamed: 0', 'STT_HO', 'STT', 'HO_TEN_CHU_HO', 'HO_TEN_THANH_VIEN',
       'QUANHE_CHUHO', 'NGAY_SINH', 'NGAY_SINH.1', 'NGAY', 'Tháng', 'Năm',
       'GIOI_TINH', 'SO_CCCD', 'TINH', 'HUYEN', 'XA', 'THÔN XÓM', 'DANTOC',
       'PHAN_LOAI_HO', 'KET_QUA', 'GHI_CHU', 'DC_DINHDANH', 'DC_CMND',
       'Unnamed: 23'],
      dtype='object')

In [273]:
df_huyen_tuy_duc.shape

(22178, 24)

- ##### Bộ dữ liệu về HUYỆN TUY ĐỨC có tổng cộng 22 181 dòng và 24 thuộc tính


#### 2.1.8. TP GIA NGHĨA

In [274]:
df_tp_gia_nghia=pd.read_excel("../data_origin/TP GIA NGHĨA.xlsx")


c:\Users\hdang\.virtualenvs\DM-xmUrTfoX\Lib\site-packages\openpyxl\worksheet\_reader.py:223: UserWarning: Cell G893 is marked as a date but the serial value 3691245 is outside the limits for dates. The cell will be treated as an error.
  warn(msg)


In [275]:
df_tp_gia_nghia.head(5)

Unnamed: 0  STT_HO  STT    HO_TEN_CHU_HO HO_TEN_THANH_VIEN QUANHE_CHUHO  \
0           1     1.0    1   NGUYỄN THỊ CÚC    NGUYỄN THỊ CÚC         HEAD   
1           2     2.0    2     LÊ THỊ THƯỢC      LÊ THỊ THƯỢC         HEAD   
2           3     3.0    3  NGUYỄN VĂN HÙNG   NGUYỄN VĂN HÙNG         HEAD   
3           4     NaN    4  NGUYỄN VĂN HÙNG   NGUYỄN THỊ HẠNH      PARTNER   
4           5     NaN    5  NGUYỄN VĂN HÙNG    NGUYỄN ANH TÀI        CHILD   

             NGAY_SINH NGAY_SINH.1  Ngày  Tháng     Năm GIOI_TINH  \
0  1948-06-10 00:00:00  10/06/1948  10.0    6.0  1948.0    FEMALE   
1  1953-03-06 00:00:00  06/03/1953   6.0    3.0  1953.0    FEMALE   
2  1958-01-01 00:00:00  01/01/1958   1.0    1.0  1958.0      MALE   
3  1959-01-01 00:00:00  01/01/1959   1.0    1.0  1959.0    FEMALE   
4  1989-10-20 00:00:00  20/10/1989  20.0   10.0  1989.0      MALE   

       SO_CCCD      TINH      HUYEN         XA Thôn, Bon, TDP DANTOC  \
0  42148007514  ĐẮK NÔNG  GIA NGHĨA  NGHĨA PHÚ          TDP 2    KIN   
1  42153001151  ĐẮK NÔNG  GIA NGHĨA  NGHĨA PHÚ          TDP 3    KIN   
2  38058026572  ĐẮK NÔNG  GIA NGHĨA  NGHĨA PHÚ          TDP 3    KIN   
3  38159013014  ĐẮK NÔNG  GIA NGHĨA  NGHĨA PHÚ          TDP 3    KIN   
4  38089023181  ĐẮK NÔNG  GIA NGHĨA  NGHĨA PHÚ          TDP 3    KIN   

  PHAN_LOAI_HO             KET_QUA GHI_CHU   DC_DINHDANH      DC_CMND  \
0         POOR  TIM THAY THEO CCCD     NaN  4.214801e+10  245023871.0   
1         POOR  TIM THAY THEO CCCD     NaN  4.215300e+10  245014782.0   
2         POOR  TIM THAY THEO CCCD     NaN  3.805803e+10  285657307.0   
3         POOR  TIM THAY THEO CCCD     NaN  3.815901e+10  285455285.0   
4         POOR  TIM THAY THEO CCCD     NaN  3.808902e+10  245372747.0   

  Unnamed: 23 Unnamed: 24  
0         NaN         NaN  
1         NaN         NaN  
2         NaN         NaN  
3         NaN         NaN  
4         NaN         NaN

In [276]:
df_tp_gia_nghia.columns

Index(['Unnamed: 0', 'STT_HO', 'STT', 'HO_TEN_CHU_HO', 'HO_TEN_THANH_VIEN',
       'QUANHE_CHUHO', 'NGAY_SINH', 'NGAY_SINH.1', 'Ngày', 'Tháng', 'Năm',
       'GIOI_TINH', 'SO_CCCD', 'TINH', 'HUYEN', 'XA', 'Thôn, Bon, TDP',
       'DANTOC', 'PHAN_LOAI_HO', 'KET_QUA', 'GHI_CHU', 'DC_DINHDANH',
       'DC_CMND', 'Unnamed: 23', 'Unnamed: 24'],
      dtype='object')

In [277]:
df_tp_gia_nghia.shape

(969, 25)

- ##### Bộ dữ liệu về TP GIA NGHĨA có tổng cộng 969 dòng và 25 thuộc tính


### 2.2.Data Pre-Processing

#### 2.2.1.Huyen_Cu_Jut

> Kiểm tra giá trị null của tập dữ liệu 

In [278]:
df_huyen_cu_jut.isnull().sum()

Unnamed: 0               0
STT_HO                9315
STT                      0
HO_TEN_CHU_HO            0
HO_TEN_THANH_VIEN        0
QUANHE\n_CHUHO           3
NGAY_SINH               10
NGAY_SINH.1           7611
Unnamed: 8            7611
Unnamed: 9            7732
Unnamed: 10           7577
Unnamed: 11          11841
GIOI_TINH                2
SO_CCCD                603
TINH                     0
HUYEN                    0
XA                       0
THÔN TỔ ẤP           11841
DAN \nTOC             4820
PHAN_\nLOAI_HO           0
KET_QUA                  0
GHI_CHU              11032
DC_DINHDANH           2873
DC_CMND               6202
Unnamed: 24          11841
Unnamed: 25          11833
dtype: int64

In [279]:
df_huyen_cu_jut.shape

(11841, 26)

> Xoá những người đã chết, chết rồi, chuyển vùng ra khỏi bộ dữ liệu

In [280]:
# List of values to filter out
values_to_drop = ["cắt hộ nghèo", "ĐÃ CHẾT", "chết", "CHẾT RỒI", "CHẾT", "CHUYỂN ĐI", "mới chết", "CHẾT RÒI", "CHUYỂN", "chuyển", "Chết", "chuyển đi", "cắt khẩu", "chuyển khẩu", "CẮT CHUYỂN KHẨU", "ĐÃ CHUYỂN HK", "KHÔNG CÓ TẠI ĐỊA PHƯƠNG", "Đã chuyển khẩu", "đã chết"]

# Dictionary to store counts
value_counts = {value: 0 for value in values_to_drop}

# Iterate over each value in values_to_drop and count occurrences
for value in values_to_drop:
    count = (df_huyen_cu_jut.isin([value])).sum().sum()
    value_counts[value] = count

# Print the counts
for value, count in value_counts.items():
    print(f"Số lượng hàng chứa '{value}': {count}")

Số lượng hàng chứa 'cắt hộ nghèo': 0
Số lượng hàng chứa 'ĐÃ CHẾT': 1
Số lượng hàng chứa 'chết': 1
Số lượng hàng chứa 'CHẾT RỒI': 2
Số lượng hàng chứa 'CHẾT': 15
Số lượng hàng chứa 'CHUYỂN ĐI': 11
Số lượng hàng chứa 'mới chết': 0
Số lượng hàng chứa 'CHẾT RÒI': 1
Số lượng hàng chứa 'CHUYỂN': 1
Số lượng hàng chứa 'chuyển': 0
Số lượng hàng chứa 'Chết': 0
Số lượng hàng chứa 'chuyển đi': 0
Số lượng hàng chứa 'cắt khẩu': 0
Số lượng hàng chứa 'chuyển khẩu': 0
Số lượng hàng chứa 'CẮT CHUYỂN KHẨU': 0
Số lượng hàng chứa 'ĐÃ CHUYỂN HK': 0
Số lượng hàng chứa 'KHÔNG CÓ TẠI ĐỊA PHƯƠNG': 0
Số lượng hàng chứa 'Đã chuyển khẩu': 0
Số lượng hàng chứa 'đã chết': 0


In [281]:

# Iterate over each column and filter rows containing values_to_drop
for column in df_huyen_cu_jut.columns:
    df_huyen_cu_jut = df_huyen_cu_jut[~df_huyen_cu_jut[column].isin(values_to_drop)]

> Quan sát thấy "Unnamed: 24","Unnamed: 25","Unnamed:  8" và "Unnamed:  9", Unnamed: 10,Unnamed: 11,NGAY_SINH.1 không có ý nghĩa phân tích vì nó chứa quá nhiều giá trị null

In [282]:
df_huyen_cu_jut.drop(columns=['Unnamed: 9','Unnamed: 8','Unnamed: 10','Unnamed: 11','NGAY_SINH.1','Unnamed: 24','Unnamed: 25'],inplace=True)

>  Kết quả sau khi xóa bỏ những dòng chứa giá trị null

In [283]:
df_huyen_cu_jut.shape

(11809, 19)

In [284]:
# List of values to filter out
values_to_drop = [
    'chết 3/4/2023',
    'chết 19/1/2023',
    'chết 17/11/2022',
    'KHÔNG CÓ TRONG HỘ KHẨU'
]

# Remove rows where 'DC_DINHDANH' contains any of the values in the list
df_huyen_cu_jut = df_huyen_cu_jut[~df_huyen_cu_jut['DC_DINHDANH'].isin(values_to_drop)]


> Các cột như Unnamed: 0 ,DC_CMND, DC_DinhDanh,Ghi_Chu,KetQua  không có ý nghĩa phân tích nên ta sẽ xoá bỏ 

In [285]:
features = ['Unnamed: 0','KET_QUA', 'GHI_CHU', 'DC_DINHDANH', 'DC_CMND','THÔN TỔ ẤP']
df_huyen_cu_jut.drop(columns=features, inplace=True)

> Kiểm tra

In [286]:
df_huyen_cu_jut.isnull().sum()

STT_HO               9287
STT                     0
HO_TEN_CHU_HO           0
HO_TEN_THANH_VIEN       0
QUANHE\n_CHUHO          3
NGAY_SINH               9
GIOI_TINH               2
SO_CCCD               598
TINH                    0
HUYEN                   0
XA                      0
DAN \nTOC            4794
PHAN_\nLOAI_HO          0
dtype: int64

> Sửa tên

In [287]:
# Đổi tên các cột
df_huyen_cu_jut = df_huyen_cu_jut.rename(columns={
    'STT_HO': 'STT_HO',
    'STT': 'STT',
    'HO_TEN_CHU_HO': 'HO_TEN_CHU_HO',
    'HO_TEN_THANH_VIEN': 'HO_TEN_THANH_VIEN',
    'QUANHE\n_CHUHO': 'QUANHE_CHUHO',
    'NGAY_SINH': 'NGAY_SINH',
    'GIOI_\nTINH': 'GIOI_TINH',
    'SO_CCCD': 'SO_CCCD',
    'TINH': 'TINH',
    'HUYEN': 'HUYEN',
    'XA': 'XA',
    'DAN \nTOC': 'DANTOC',
    'PHAN_\nLOAI_HO': 'PHAN_LOAI_HO'
})
df_huyen_cu_jut.columns


Index(['STT_HO', 'STT', 'HO_TEN_CHU_HO', 'HO_TEN_THANH_VIEN', 'QUANHE_CHUHO',
       'NGAY_SINH', 'GIOI_TINH', 'SO_CCCD', 'TINH', 'HUYEN', 'XA', 'DANTOC',
       'PHAN_LOAI_HO'],
      dtype='object')

> Kiểm tra

In [288]:
df_huyen_cu_jut.isnull().sum()

STT_HO               9287
STT                     0
HO_TEN_CHU_HO           0
HO_TEN_THANH_VIEN       0
QUANHE_CHUHO            3
NGAY_SINH               9
GIOI_TINH               2
SO_CCCD               598
TINH                    0
HUYEN                   0
XA                      0
DANTOC               4794
PHAN_LOAI_HO            0
dtype: int64

In [289]:
df_huyen_cu_jut.shape

(11804, 13)

> Xem các dòng mang giá trị bị thiếu của cột giới tính

In [290]:
rows_with_null_gioi_tinh=df_huyen_cu_jut[df_huyen_cu_jut['GIOI_TINH'].isnull()]
print(rows_with_null_gioi_tinh)

       STT_HO    STT   HO_TEN_CHU_HO HO_TEN_THANH_VIEN  QUANHE_CHUHO  \
10859     2.0  10860    H' ĐLIU ÊBAN      H' ĐLIU ÊBAN           1.0   
10865     NaN  10866  Y BLIỄU BKRÔNG       Y HÙNG ÊBAN           3.0   

        NGAY_SINH  GIOI_TINH       SO_CCCD      TINH   HUYEN        XA DANTOC  \
10859  01/01/1957        NaN  067157000089  ĐẮK NÔNG  CƯ JUT  NAM DONG     12   
10865  01/08/2009        NaN  067209008290  ĐẮK NÔNG  CƯ JUT  NAM DONG     12   

       PHAN_LOAI_HO  
10859             2  
10865             2  


> Thường thì họ "Y" thường là "Nam" điền "1", ở Người Ê Đê, và họ 'H' thường là "Nữ" điền "2"

In [291]:
# Điền giá trị vào cột 'GIOI_TINH' cho người có 'SO_CCCD' là '067157000089'
df_huyen_cu_jut.loc[df_huyen_cu_jut['SO_CCCD'] == '067157000089', 'GIOI_TINH'] = 2

# Điền giá trị vào cột 'GIOI_TINH' cho người có 'SO_CCCD' là '067209008290'
df_huyen_cu_jut.loc[df_huyen_cu_jut['SO_CCCD'] == '067209008290', 'GIOI_TINH'] = 1

> Xem các dòng mang giá trị bị thiếu của cột  NGAY_SINH

In [292]:
rows_with_null_ngay_sinh=df_huyen_cu_jut[df_huyen_cu_jut['NGAY_SINH'].isnull()]
print(rows_with_null_ngay_sinh[['HO_TEN_THANH_VIEN','SO_CCCD','XA']])

         HO_TEN_THANH_VIEN       SO_CCCD         XA
4798        BÙI TRỌNG TUẤN  067219003315   NAM DONG
5697        NGUYỄN VĂN BẮC  036202014319  TÂM THẮNG
6199        H NAHI BKRRÔNG           NaN  TÂM THẮNG
8981           ĐỖ HOA THÊU  036171019356   CƯ K'NIA
8984     VĨNH THỊ ÚT QUANG  067193003842   CƯ K'NIA
8985          ĐẶNG THỊ SỬU  036149007005   CƯ K'NIA
8986       HOÀNG THỦY TIÊN  067318006292   CƯ K'NIA
8993   TRẦN THỊ ĐOAN TRANG  067312001319   CƯ K'NIA
11623        HÀ VĂN THƯƠNG  038088035385      EA PÔ


>Tính tuổi đối với những hàng nào tính độ tuổi không được thì sẽ điền giá trị trung bình

In [293]:
def calculate_age(born):
    today = datetime.today().date()  # Lấy ngày hiện tại mà không có giờ, phút, giây
    try:
        # Xử lý các định dạng không hợp lệ
        if "-" in born:  # Nếu có dấu "-"
            parts = born.split("-")
            if len(parts[0]) == 4:  # Nếu định dạng là yyyy-mm-dd
                born = datetime.strptime(born, "%Y-%m-%d").date()
            else:  # Nếu định dạng là dd-mm-yyyy
                born = datetime.strptime(born, "%d-%m-%Y").date()
        elif "/" in born:  # Nếu có dấu "/"
            born = datetime.strptime(born, "%d/%m/%Y").date()
        else:
            # Giả định ngày sinh là ngày 1 tháng 1 của năm đó
            born = datetime(int(born), 1, 1).date()
        
        return today.year - born.year
    except:
        return None

# Áp dụng hàm tính tuổi vào cột 'NGAY_SINH'
df_huyen_cu_jut['TUOI'] = df_huyen_cu_jut['NGAY_SINH'].apply(calculate_age)

> Điền giá trị trung bình cho cột tuổi của những hàng nào không tính được

In [294]:
df_huyen_cu_jut.isnull().sum()

STT_HO               9287
STT                     0
HO_TEN_CHU_HO           0
HO_TEN_THANH_VIEN       0
QUANHE_CHUHO            3
NGAY_SINH               9
GIOI_TINH               0
SO_CCCD               598
TINH                    0
HUYEN                   0
XA                      0
DANTOC               4794
PHAN_LOAI_HO            0
TUOI                   41
dtype: int64

In [295]:
df_huyen_cu_jut.shape

(11804, 14)

In [296]:
# Tính giá trị trung bình  trong cột Tuổi
most_common_year_old =round(df_huyen_cu_jut['TUOI'].mean())

# Điền giá trị trung bình vào các ô trống
df_huyen_cu_jut.fillna({'TUOI': most_common_year_old }, inplace=True)

In [297]:
df_huyen_cu_jut.isnull().sum()

STT_HO               9287
STT                     0
HO_TEN_CHU_HO           0
HO_TEN_THANH_VIEN       0
QUANHE_CHUHO            3
NGAY_SINH               9
GIOI_TINH               0
SO_CCCD               598
TINH                    0
HUYEN                   0
XA                      0
DANTOC               4794
PHAN_LOAI_HO            0
TUOI                    0
dtype: int64

In [298]:
df_huyen_cu_jut.head(5)

STT_HO  STT     HO_TEN_CHU_HO  HO_TEN_THANH_VIEN  QUANHE_CHUHO   NGAY_SINH  \
0     1.0    1    TRƯƠNG THỊ LÙN     TRƯƠNG THỊ LÙN           1.0  22/04/1905   
1     2.0    2  HOÀNG VĂN TRƯỜNG   HOÀNG VĂN TRƯỜNG           1.0  02/06/1905   
2     NaN    3  HOÀNG VĂN TRƯỜNG       PHẠM THỊ SOI           2.0  03/06/1905   
3     NaN    4  HOÀNG VĂN TRƯỜNG  HOÀNG THỊ BẢO THI           3.0  03/07/1905   
4     NaN    5  HOÀNG VĂN TRƯỜNG       HOÀNG KỲ ANH           3.0  05/07/1905   

   GIOI_TINH       SO_CCCD      TINH   HUYEN         XA DANTOC  PHAN_LOAI_HO  \
0        2.0  054139002355  ĐẮK NÔNG  CƯ JUT  TÂM THẮNG      1             1   
1        1.0  036080016941  ĐẮK NÔNG  CƯ JUT  TÂM THẮNG      1             1   
2        2.0  036181021986  ĐẮK NÔNG  CƯ JUT  TÂM THẮNG      1             1   
3        2.0  067311001529  ĐẮK NÔNG  CƯ JUT  TÂM THẮNG      1             1   
4        1.0  067213007898  ĐẮK NÔNG  CƯ JUT  TÂM THẮNG      1             1   

    TUOI  
0  119.0  
1  119.0  
2  119.0  
3  119.0  
4  119.0

> Điền giá trị phổ biến cho cột Dân Tộc theo từng xã.Trường hợp nếu toàn bộ xã đó ở cột dân tộc các hàng đều mang giá trị bị thiếu thì ta điền giá trị phổ biến

In [299]:
def fill_missing_dantoc(df):
    # Tìm giá trị phổ biến của toàn bộ dataframe
    overall_most_frequent_dantoc = df['DANTOC'].mode().iloc[0]

    for xa in df['XA'].unique():
        dantoc_series = df[df['XA'] == xa]['DANTOC'].mode()
        if not dantoc_series.empty:
            most_frequent_dantoc = dantoc_series.iloc[0]
        else:
            most_frequent_dantoc = overall_most_frequent_dantoc
        
        df.loc[df['XA'] == xa, 'DANTOC'] = df.loc[df['XA'] == xa, 'DANTOC'].fillna(most_frequent_dantoc)
    
    return df


In [300]:
df_huyen_cu_jut= fill_missing_dantoc(df_huyen_cu_jut) 

In [301]:
df_huyen_cu_jut.isnull().sum()

STT_HO               9287
STT                     0
HO_TEN_CHU_HO           0
HO_TEN_THANH_VIEN       0
QUANHE_CHUHO            3
NGAY_SINH               9
GIOI_TINH               0
SO_CCCD               598
TINH                    0
HUYEN                   0
XA                      0
DANTOC                  0
PHAN_LOAI_HO            0
TUOI                    0
dtype: int64

In [302]:
df_huyen_cu_jut.columns

Index(['STT_HO', 'STT', 'HO_TEN_CHU_HO', 'HO_TEN_THANH_VIEN', 'QUANHE_CHUHO',
       'NGAY_SINH', 'GIOI_TINH', 'SO_CCCD', 'TINH', 'HUYEN', 'XA', 'DANTOC',
       'PHAN_LOAI_HO', 'TUOI'],
      dtype='object')

> Điền giá trị phổ biến cho cột QUANHE_CHUHO theo từng xã

In [303]:
def fill_missing_quanhe_chuho(df):
    # Tìm giá trị phổ biến của toàn bộ dataframe
    overall_most_frequent_quanhe_chuho = df['QUANHE_CHUHO'].mode().iloc[0]

    for xa in df['XA'].unique():
        quanhe_chuho_series = df[df['XA'] == xa]['QUANHE_CHUHO'].mode()
        if not quanhe_chuho_series.empty:
            most_frequent_quanhe_chuho = quanhe_chuho_series.iloc[0]
        else:
            most_frequent_quanhe_chuho = overall_most_frequent_quanhe_chuho
        
        df.loc[df['XA'] == xa, 'QUANHE_CHUHO'] = df.loc[df['XA'] == xa, 'QUANHE_CHUHO'].fillna(most_frequent_quanhe_chuho)
    
    return df


In [304]:
df_huyen_cu_jut = fill_missing_quanhe_chuho(df_huyen_cu_jut)

In [305]:
df_huyen_cu_jut.isnull().sum()

STT_HO               9287
STT                     0
HO_TEN_CHU_HO           0
HO_TEN_THANH_VIEN       0
QUANHE_CHUHO            0
NGAY_SINH               9
GIOI_TINH               0
SO_CCCD               598
TINH                    0
HUYEN                   0
XA                      0
DANTOC                  0
PHAN_LOAI_HO            0
TUOI                    0
dtype: int64

> Xóa dòng trùng lặp

In [306]:
df_huyen_cu_jut.drop_duplicates(inplace=True)

> Sau khi tiền xử lí thì từ (11843, 26) gốc còn lại (11805, 14)

In [307]:
df_huyen_cu_jut.shape

(11804, 14)

> Xuất file

In [308]:
df_huyen_cu_jut.shape

(11804, 14)

In [309]:
df_huyen_cu_jut.to_excel("../data_processed/HUYỆN CƯ JUT.xlsx")

#### 2.2.2. Huyện Dak Glong

> Kiểm tra giá trị null của tập dữ liệu 

In [310]:
df_huyen_dak_glong.isnull().sum()

Unnamed: 0               0
STT_HO               30372
STT                      0
HO_TEN_CHU_HO            0
HO_TEN_THANH_VIEN        0
QUANHE_CHUHO           114
NGAY_SINH               37
Unnamed: 7           37427
Unnamed: 8           37427
Unnamed: 9           37427
Unnamed: 10          37427
Unnamed: 11          37427
GIOI_TINH             1292
SO_CCCD               4610
TINH                     0
HUYEN                    0
XA                       0
DANTOC                  57
PHAN_LOAI_HO             0
KET_QUA                  0
DC_DINHDANH           3955
DC_CMND              20487
GHI_CHU              33127
dtype: int64

In [311]:
df_huyen_dak_glong.shape

(37427, 23)

> Xoá những người chuyển vùng,chết rồi,đã chết 

In [312]:
# List of values to filter out
values_to_drop = ["cắt hộ nghèo","ĐÃ CHẾT", "chết", "CHẾT RỒI","CHẾT", "CHUYỂN ĐI","mới chết", "CHẾT RÒI", "CHUYỂN","chuyển","Chết","chuyển đi","cắt khẩu","chuyển khẩu","CẮT CHUYỂN KHẨU","ĐÃ CHUYỂN HK","KHÔNG CÓ TẠI ĐỊA PHƯƠNG","Đã chuyển khẩu","đã chết"]
# Iterate over each column and filter rows containing values_to_drop
for column in df_huyen_dak_glong.columns:
    df_huyen_dak_glong = df_huyen_dak_glong[~df_huyen_dak_glong[column].isin(values_to_drop)]

> Các cột như Unnamed: 7 ,Unnamed: 8, Unnamed: 9, Unnamed: 10,Unnamed: 11 chứa quá nhiều giá trị null cho nên ta sẽ xoá bỏ ra khỏi bộ dữ liệu

In [313]:
df_huyen_dak_glong.shape

(37426, 23)

In [314]:
df_huyen_dak_glong.drop(columns=['Unnamed: 7','Unnamed: 8', 'Unnamed: 9',
       'Unnamed: 10', 'Unnamed: 11'],inplace=True)

> Kết quả sau khi xóa bỏ

In [315]:
df_huyen_dak_glong.shape

(37426, 18)

In [316]:
df_huyen_dak_glong.columns

Index(['Unnamed: 0', 'STT_HO', 'STT', 'HO_TEN_CHU_HO', 'HO_TEN_THANH_VIEN',
       'QUANHE_CHUHO', 'NGAY_SINH', 'GIOI_TINH', 'SO_CCCD', 'TINH', 'HUYEN',
       'XA', 'DANTOC', 'PHAN_LOAI_HO', 'KET_QUA', 'DC_DINHDANH', 'DC_CMND',
       'GHI_CHU'],
      dtype='object')

> Các cột như DC_CMND, DC_DinhDanh,Ghi_Chu,KetQua,Unnamed: 0 không có ý nghĩa phân tích nên ta sẽ xoá bỏ 

In [317]:
feature = ['KET_QUA', 'GHI_CHU', 'DC_DINHDANH', 'DC_CMND', 'Unnamed: 0']
df_huyen_dak_glong.drop(columns=feature, inplace=True)

In [318]:
df_huyen_dak_glong.shape

(37426, 13)

> Xóa giá trị trùng lặp 

In [319]:
df_huyen_dak_glong.drop_duplicates(inplace=True)

In [320]:
df_huyen_dak_glong.isnull().sum()

STT_HO               30371
STT                      0
HO_TEN_CHU_HO            0
HO_TEN_THANH_VIEN        0
QUANHE_CHUHO           114
NGAY_SINH               37
GIOI_TINH             1292
SO_CCCD               4610
TINH                     0
HUYEN                    0
XA                       0
DANTOC                  57
PHAN_LOAI_HO             0
dtype: int64

In [321]:
df_huyen_dak_glong.shape

(37426, 13)

In [322]:
df_huyen_dak_glong['XA'].unique()

array(['ĐẮK PLAO', 'ĐẮK HA', "ĐẮK R'MĂNG", 'QUẢNG KHÊ', 'QUẢNG HÒA',
       'QUẢNG SƠN', 'ĐẮK SOM'], dtype=object)

> Điền giá trị phổ biến cho cột dân tộc theo từng xã

In [323]:
df_huyen_dak_glong=fill_missing_dantoc(df_huyen_dak_glong)

In [324]:
df_huyen_dak_glong.isnull().sum()

STT_HO               30371
STT                      0
HO_TEN_CHU_HO            0
HO_TEN_THANH_VIEN        0
QUANHE_CHUHO           114
NGAY_SINH               37
GIOI_TINH             1292
SO_CCCD               4610
TINH                     0
HUYEN                    0
XA                       0
DANTOC                   0
PHAN_LOAI_HO             0
dtype: int64

> Tính tuổi dựa trên cột ngày sinh

In [325]:
df_huyen_dak_glong['TUOI'] = df_huyen_dak_glong['NGAY_SINH'].apply(calculate_age)

In [326]:
df_huyen_dak_glong.isnull().sum()

STT_HO               30371
STT                      0
HO_TEN_CHU_HO            0
HO_TEN_THANH_VIEN        0
QUANHE_CHUHO           114
NGAY_SINH               37
GIOI_TINH             1292
SO_CCCD               4610
TINH                     0
HUYEN                    0
XA                       0
DANTOC                   0
PHAN_LOAI_HO             0
TUOI                   222
dtype: int64

> Đối với hàng nào tính không được ta sẽ điền giá trị trung bình của tuổi

In [327]:
# Tính giá trị trung bình  trong cột Tuổi
most_common_year_old = round(df_huyen_dak_glong['TUOI'].mean())

# Điền giá trị trung bình  vào các ô trống
df_huyen_dak_glong.fillna({'TUOI': most_common_year_old }, inplace=True)

> Điền giá trị phổ biến cho  cột giới tính

In [328]:
# Tính giá trị phổ biến nhất trong cột GIOI_TINH
most_common_gender = df_huyen_dak_glong['GIOI_TINH'].mode()[0]

# Điền giá trị phổ biến vào các ô trống
df_huyen_dak_glong.fillna({'GIOI_TINH': most_common_gender}, inplace=True)


> Điền giá trị phổ biến cho cột QUANHE_CHUHO theo từng xã

In [329]:
df_huyen_dak_glong = fill_missing_quanhe_chuho(df_huyen_dak_glong)

In [330]:
df_huyen_dak_glong.isnull().sum()

STT_HO               30371
STT                      0
HO_TEN_CHU_HO            0
HO_TEN_THANH_VIEN        0
QUANHE_CHUHO             0
NGAY_SINH               37
GIOI_TINH                0
SO_CCCD               4610
TINH                     0
HUYEN                    0
XA                       0
DANTOC                   0
PHAN_LOAI_HO             0
TUOI                     0
dtype: int64

> Ban đầu có (37427, 23) sau quá trình tiền xử lí dữ liêu thì còn (32817, 14)

In [331]:
df_huyen_dak_glong.shape

(37426, 14)

In [332]:
df_huyen_dak_glong.shape

(37426, 14)

> Xuất File

In [333]:
df_huyen_dak_glong.to_excel("../data_processed/Huyện Đắk Glong.xlsx")

#### 2.2.3. HUYỆN DAK MIL.xlsx

>Kiểm tra giá trị null của tập dữ liệu 

In [334]:
df_huyen_dak_mil.isnull().sum()

NAM RA SOAT             0
STT_HO               3709
STT                     0
LOẠI HỘ                 0
Khu vực                 0
MÃ THÔN                 0
THÔN                    0
HO_TEN_CHU_HO         527
HO_TEN_THANH_VIEN       0
SO_CCCD               388
QUANHE_CHUHO            0
NGAY_SINH             668
Tháng                 547
Ngày                  539
DANTOC                196
GIOI_TINH              62
TINH                    0
PHAN_LOAI_HO            0
HUYEN                   0
XA                      0
dtype: int64

In [335]:
df_huyen_dak_mil.shape

(4723, 20)

> Xoá những người đã chết, chết rồi,chuyển vùng

In [336]:
# List of values to filter out
values_to_drop = ["cắt hộ nghèo","ĐÃ CHẾT", "chết", "CHẾT RỒI","CHẾT", "CHUYỂN ĐI","mới chết", "CHẾT RÒI", "CHUYỂN","chuyển","Chết","chuyển đi","cắt khẩu","chuyển khẩu","CẮT CHUYỂN KHẨU","ĐÃ CHUYỂN HK","KHÔNG CÓ TẠI ĐỊA PHƯƠNG","Đã chuyển khẩu","đã chết"]
# Iterate over each column and filter rows containing values_to_drop
for column in df_huyen_dak_mil.columns:
    df_huyen_dak_mil = df_huyen_dak_mil[~df_huyen_dak_mil[column].isin(values_to_drop)]

> Kết quả sau khi loại bỏ

In [337]:
df_huyen_dak_mil.shape

(4722, 20)

In [338]:
df_huyen_dak_mil.columns

Index(['NAM RA SOAT', 'STT_HO', 'STT', 'LOẠI HỘ ', 'Khu vực', 'MÃ THÔN',
       'THÔN', 'HO_TEN_CHU_HO', 'HO_TEN_THANH_VIEN', 'SO_CCCD', 'QUANHE_CHUHO',
       'NGAY_SINH', 'Tháng', 'Ngày', 'DANTOC', 'GIOI_TINH', 'TINH',
       'PHAN_LOAI_HO', 'HUYEN', 'XA'],
      dtype='object')

> Các cột như 'NAM RA SOAT','LOẠI HỘ ', 'Khu vực', 'MÃ THÔN', 'THÔN','Tháng', 'Ngày',không có ý nghĩa phân tích nên ta sẽ xoá bỏ 

In [339]:
features=['NAM RA SOAT', 'LOẠI HỘ ', 'Khu vực', 'MÃ THÔN', 'THÔN','Tháng', 'Ngày']
df_huyen_dak_mil.drop(columns=features,inplace=True)

In [340]:
df_huyen_dak_mil.shape

(4722, 13)

In [341]:
df_huyen_dak_mil.columns

Index(['STT_HO', 'STT', 'HO_TEN_CHU_HO', 'HO_TEN_THANH_VIEN', 'SO_CCCD',
       'QUANHE_CHUHO', 'NGAY_SINH', 'DANTOC', 'GIOI_TINH', 'TINH',
       'PHAN_LOAI_HO', 'HUYEN', 'XA'],
      dtype='object')

> Đối với dữ liệu ở cột PHÂN LOẠI HỘ ta thực hiện chuẩn hoá thành 1 : POOR, 2 : NEAR_POOR

In [342]:
df_huyen_dak_mil['PHAN_LOAI_HO'].unique()

array(['POOR', 'NEAR_POOR'], dtype=object)

In [343]:
df_huyen_dak_mil['PHAN_LOAI_HO']=df_huyen_dak_mil['PHAN_LOAI_HO'].replace({1:'POOR',2:'NEAR_POOR'})

In [344]:
df_huyen_dak_mil.shape

(4722, 13)

> Định dạng lại thứ tự 

In [345]:
df_huyen_dak_mil = df_huyen_dak_mil[['STT_HO', 'STT', 'HO_TEN_CHU_HO', 'HO_TEN_THANH_VIEN', 'QUANHE_CHUHO', 
         'NGAY_SINH', 'GIOI_TINH', 'SO_CCCD', 'TINH', 'HUYEN', 'XA', 'DANTOC', 'PHAN_LOAI_HO']]

> Xóa giá trị trùng lặp

In [346]:
df_huyen_dak_mil.drop_duplicates(inplace=True)

In [347]:
df_huyen_dak_mil.isnull().sum()

STT_HO               3709
STT                     0
HO_TEN_CHU_HO         527
HO_TEN_THANH_VIEN       0
QUANHE_CHUHO            0
NGAY_SINH             668
GIOI_TINH              62
SO_CCCD               388
TINH                    0
HUYEN                   0
XA                      0
DANTOC                196
PHAN_LOAI_HO            0
dtype: int64

> Điền giá trị phổ biến cho cột dân tộc theo từng xã

In [348]:
df_huyen_dak_mil=fill_missing_dantoc(df_huyen_dak_mil)

> Tính tuổi dựa trên ngày sinh

In [349]:
def calculate_age_from_year(birth_year):
    try:
        current_year = datetime.now().year
        age = current_year - birth_year
        return age
    except Exception as e:
        print(f"Error processing birth year {birth_year}: {e}")
        return None

In [350]:
df_huyen_dak_mil['TUOI'] =round(df_huyen_dak_mil['NGAY_SINH'].apply(calculate_age_from_year))

In [351]:
df_huyen_dak_mil.isnull().sum()

STT_HO               3709
STT                     0
HO_TEN_CHU_HO         527
HO_TEN_THANH_VIEN       0
QUANHE_CHUHO            0
NGAY_SINH             668
GIOI_TINH              62
SO_CCCD               388
TINH                    0
HUYEN                   0
XA                      0
DANTOC                  0
PHAN_LOAI_HO            0
TUOI                  668
dtype: int64

> Điền giá trị phổ biến cho những hàng nào tính không được

In [352]:
# Tính giá trị phổ biến nhất trong cột TUOI
most_common_age = df_huyen_dak_mil['TUOI'].mode()[0]

# Điền giá trị phổ biến nhất vào các ô trống
df_huyen_dak_mil['TUOI'].fillna(most_common_age, inplace=True)

C:\Users\hdang\AppData\Local\Temp\ipykernel_11408\20731481.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_huyen_dak_mil['TUOI'].fillna(most_common_age, inplace=True)


> Điền giá trị phổ biến cho cột giới tính

In [353]:
most_common_gender=df_huyen_dak_mil['GIOI_TINH'].mode().iloc[0]
df_huyen_dak_mil.fillna({'GIOI_TINH':most_common_gender},inplace=True)

>Điền giá trị phổ biến cho cột QUANHE_CHUHO theo từng xã

In [354]:
df_huyen_dak_mil = fill_missing_quanhe_chuho(df_huyen_dak_mil)

> Ban đầu có (4724, 20) sau khi tiền xử lí thì còn lại (4335, 14)

In [355]:
df_huyen_dak_mil.shape

(4722, 14)

In [356]:
df_huyen_dak_mil["XA"].unique()

array(['Thị trấn Đắk Mil', 'Đắk Gằn', 'Đắk Lao', "ĐẮK R'LA", 'ĐỨC MẠNH',
       'THUẬN AN', 'ĐẮk SẮK', 'ĐỨC MINH', 'LONG SƠN'], dtype=object)

In [357]:
df_huyen_dak_mil.isnull().sum()

STT_HO               3709
STT                     0
HO_TEN_CHU_HO         527
HO_TEN_THANH_VIEN       0
QUANHE_CHUHO            0
NGAY_SINH             668
GIOI_TINH               0
SO_CCCD               388
TINH                    0
HUYEN                   0
XA                      0
DANTOC                  0
PHAN_LOAI_HO            0
TUOI                    0
dtype: int64

In [358]:
df_huyen_dak_mil['DANTOC'].unique()

array(['KIN', 'NUN', 'TAY', 'MNO', 'MUO', 'GIR', 'THAI', 'KHM', 'DAO',
       'XED', 'THA', 'HMO', 'CHA'], dtype=object)

> Chuyển đổi dữ liệu sang biến phân loại ở cột "DANTOC","PHAN_LOAI_HO","GIOI_TINH"

In [359]:
dantoc_mapping = {
    'KIN': 1,
    'NUN': 7,
    'TAY': 2,
    'MNO': 20,
    'MUO': 6,
    'GIR': 10,
    'THAI': 3,
    'KHM': 5,
    'DAO': 9,
    'THA': 3,
    'HMO': 8,
    'CHA':17,
    'XED':14
}
# Thay đổi giá trị trong cột 'DANTOC' theo từ điển ánh xạ
df_huyen_dak_mil['DANTOC'] = df_huyen_dak_mil['DANTOC'].map(dantoc_mapping)

In [360]:
gioi_tinh_mapping = {
    'FEMALE': 2,
    'MALE': 1
}

# Thay đổi giá trị trong cột 'GIOI_TINH' theo từ điển ánh xạ
df_huyen_dak_mil['GIOI_TINH'] = df_huyen_dak_mil['GIOI_TINH'].map(gioi_tinh_mapping)

In [361]:
df_huyen_dak_mil['PHAN_LOAI_HO'].unique()

array(['POOR', 'NEAR_POOR'], dtype=object)

In [362]:
phan_loai_ho_mapping={
    'POOR':1,
    'NEAR_POOR':2
}
df_huyen_dak_mil['PHAN_LOAI_HO'] = df_huyen_dak_mil['PHAN_LOAI_HO'].map(phan_loai_ho_mapping)

In [363]:
df_huyen_dak_mil.isnull().sum()

STT_HO               3709
STT                     0
HO_TEN_CHU_HO         527
HO_TEN_THANH_VIEN       0
QUANHE_CHUHO            0
NGAY_SINH             668
GIOI_TINH               0
SO_CCCD               388
TINH                    0
HUYEN                   0
XA                      0
DANTOC                  0
PHAN_LOAI_HO            0
TUOI                    0
dtype: int64

> Xuất file

In [364]:
df_huyen_dak_mil.to_excel("../data_processed/HUYỆN DAK MIL.xlsx")

#### 2.2.4. HUYỆN ĐẮK RLẤP

>Kiểm tra giá trị null của tập dữ liệu 

In [365]:
df_huyen_dak_rlap.isnull().sum()

STT_HO               3177
STT                     0
HO_TEN_CHU_HO           0
HO_TEN_THANH_VIEN       2
QUANHE_CHUHO            3
NGAY_SINH              17
GIOI_TINH              13
SO_CCCD                40
TINH                    0
HUYEN                   0
XA                      0
DANTOC                  5
PHAN_LOAI_HO            0
KET_QUA                 3
GHI_CHU              3829
DC_DINHDANH           652
DC_CMND              1956
GHI CHÚ              4193
dtype: int64

In [366]:
df_huyen_dak_rlap.shape

(4241, 18)

In [367]:
df_huyen_dak_rlap.columns

Index(['STT_HO', 'STT', 'HO_TEN_CHU_HO', 'HO_TEN_THANH_VIEN', 'QUANHE_CHUHO',
       'NGAY_SINH', 'GIOI_TINH', 'SO_CCCD', 'TINH', 'HUYEN', 'XA', 'DANTOC',
       'PHAN_LOAI_HO', 'KET_QUA', 'GHI_CHU', 'DC_DINHDANH', 'DC_CMND',
       'GHI CHÚ '],
      dtype='object')

> Xoá những người đã chết,chết,chuyển vùng ra khỏi bộ dữ liệu

In [368]:
# List of values to filter out
values_to_drop = ["cắt hộ nghèo","ĐÃ CHẾT", "chết", "CHẾT RỒI","CHẾT", "CHUYỂN ĐI","mới chết", "CHẾT RÒI", "CHUYỂN","chuyển","Chết","chuyển đi","cắt khẩu","chuyển khẩu","CẮT CHUYỂN KHẨU","ĐÃ CHUYỂN HK","KHÔNG CÓ TẠI ĐỊA PHƯƠNG","Đã chuyển khẩu","đã chết"]

# Iterate over each column and filter rows containing values_to_drop
for column in df_huyen_dak_rlap.columns:
    df_huyen_dak_rlap = df_huyen_dak_rlap[~df_huyen_dak_rlap[column].isin(values_to_drop)]

> Kết quả sau khi loại bỏ

In [369]:
df_huyen_dak_rlap.shape

(4220, 18)

In [370]:
df_huyen_dak_rlap.columns

Index(['STT_HO', 'STT', 'HO_TEN_CHU_HO', 'HO_TEN_THANH_VIEN', 'QUANHE_CHUHO',
       'NGAY_SINH', 'GIOI_TINH', 'SO_CCCD', 'TINH', 'HUYEN', 'XA', 'DANTOC',
       'PHAN_LOAI_HO', 'KET_QUA', 'GHI_CHU', 'DC_DINHDANH', 'DC_CMND',
       'GHI CHÚ '],
      dtype='object')

> Các cột như DC_CMND, DC_DinhDanh,Ghi_Chu,KetQua,Unnamed: 2 không có ý nghĩa phân tích nên ta sẽ xoá bỏ 

In [371]:
features=['KET_QUA','GHI_CHU', 'DC_DINHDANH', 'DC_CMND', 'GHI CHÚ ']
df_huyen_dak_rlap.drop(columns=features,inplace=True)

In [372]:
df_huyen_dak_rlap.shape

(4220, 13)

In [373]:
df_huyen_dak_rlap.columns

Index(['STT_HO', 'STT', 'HO_TEN_CHU_HO', 'HO_TEN_THANH_VIEN', 'QUANHE_CHUHO',
       'NGAY_SINH', 'GIOI_TINH', 'SO_CCCD', 'TINH', 'HUYEN', 'XA', 'DANTOC',
       'PHAN_LOAI_HO'],
      dtype='object')

> Xóa giá trị trùng lặp

In [374]:
df_huyen_dak_rlap.drop_duplicates(inplace=True)

In [375]:
df_huyen_dak_rlap.shape

(4220, 13)

In [376]:
df_huyen_dak_rlap.isnull().sum()

STT_HO               3164
STT                     0
HO_TEN_CHU_HO           0
HO_TEN_THANH_VIEN       2
QUANHE_CHUHO            3
NGAY_SINH              16
GIOI_TINH              13
SO_CCCD                39
TINH                    0
HUYEN                   0
XA                      0
DANTOC                  5
PHAN_LOAI_HO            0
dtype: int64

> Điền giá trị phổ biến cho cột Dân Tộc theo xã

In [377]:
df_huyen_dak_rlap=fill_missing_dantoc(df_huyen_dak_rlap)

> Điền giá trị phổ biến cho cột giới tính

In [378]:
# Tính giá trị phổ biến nhất trong cột GIOI_TINH
most_common_gender = df_huyen_dak_rlap['GIOI_TINH'].mode()[0]

# Điền giá trị phổ biến vào các ô trống
df_huyen_dak_rlap.fillna({'GIOI_TINH': most_common_gender}, inplace=True)



> Tính tuổi dựa trên ngày sinh

In [379]:
df_huyen_dak_rlap['TUOI']=round(df_huyen_dak_rlap['NGAY_SINH'].apply(calculate_age))

In [380]:
df_huyen_dak_rlap.isnull().sum()

STT_HO               3164
STT                     0
HO_TEN_CHU_HO           0
HO_TEN_THANH_VIEN       2
QUANHE_CHUHO            3
NGAY_SINH              16
GIOI_TINH               0
SO_CCCD                39
TINH                    0
HUYEN                   0
XA                      0
DANTOC                  0
PHAN_LOAI_HO            0
TUOI                   89
dtype: int64

> Điền giá trị trung bình cho hàng nào không tính được ở cột tuổi

In [381]:
# Tính giá trị trung bình trong cột Tuổi
most_common_year_old = df_huyen_dak_rlap['TUOI'].mean()

# Điền giá trị trung bình  vào các ô trống
df_huyen_dak_rlap.fillna({'TUOI': most_common_year_old }, inplace=True)


> Điền giá trị phổ biến cho cột QUANHE_CHUHO theo từng xã

In [382]:
df_huyen_dak_rlap = fill_missing_quanhe_chuho(df_huyen_dak_rlap)

> Ban đầu ta có (4241, 18) sau khi tiền xử lí thì còn lại (4201, 14)

In [383]:
df_huyen_dak_rlap.shape

(4220, 14)

In [384]:
df_huyen_dak_rlap.isnull().sum()

STT_HO               3164
STT                     0
HO_TEN_CHU_HO           0
HO_TEN_THANH_VIEN       2
QUANHE_CHUHO            0
NGAY_SINH              16
GIOI_TINH               0
SO_CCCD                39
TINH                    0
HUYEN                   0
XA                      0
DANTOC                  0
PHAN_LOAI_HO            0
TUOI                    0
dtype: int64

In [385]:
df_huyen_dak_rlap['XA'].unique()

array(['NHÂN CƠ', 'KIẾN ĐỨC', 'QUẢNG TÍN', 'ĐẠO NGHĨA', 'KIẾN THÀNH',
       'ĐĂK SIN', 'HƯNG BÌNH', 'NHÂN ĐẠO', 'ĐĂK RU', 'NGHĨA THẮNG',
       'ĐẮK WER'], dtype=object)

In [386]:
df_huyen_dak_rlap['GIOI_TINH'].unique()

array([1.00000000e+00, 2.00000000e+00, 4.20650005e+11, 4.21320059e+10])

> xoá các giá trị trong cột giới tính mang giá trị không đúng định dạng

In [387]:
# Giá trị cần xoá
values_to_remove = ['420650005314.0', '42132005926.0']

# Chuyển đổi cột 'GIOI_TINH' sang chuỗi nếu cần thiết
df_huyen_dak_rlap['GIOI_TINH'] = df_huyen_dak_rlap['GIOI_TINH'].astype(str)

# Lọc và gán lại DataFrame
df_huyen_dak_rlap = df_huyen_dak_rlap[~df_huyen_dak_rlap['GIOI_TINH'].isin(values_to_remove)]

> Xuất file

In [388]:
df_huyen_dak_rlap.to_excel("../data_processed/HUYỆN ĐẮK RLẤP.xlsx")

#### 2.2.5. HUYỆN ĐẮK SONG

> Kiểm tra giá trị null

In [389]:
df_huyen_dak_song.isnull().sum()

Unnamed: 0              0
STT_HO               7953
STT                  2201
HO_TEN_CHU_HO           0
HO_TEN_THANH_VIEN       1
QUANHE_CHUHO            3
NGAY_SINH               1
NGAY_SINH.1             1
Ngay                  289
Tháng                 289
Năm                    48
GIOI_TINH             480
SO_CCCD               410
TINH                    0
HUYEN                   0
XA                      0
DANTOC                809
PHAN_LOAI_HO            0
KET_QUA                33
GHI_CHU              8361
DC_DINHDANH          2300
DC_CMND              5130
dtype: int64

In [390]:
df_huyen_dak_song.shape

(9028, 22)

In [391]:
df_huyen_dak_song.columns

Index(['Unnamed: 0', 'STT_HO', 'STT', 'HO_TEN_CHU_HO', 'HO_TEN_THANH_VIEN',
       'QUANHE_CHUHO', 'NGAY_SINH', 'NGAY_SINH.1', 'Ngay', 'Tháng', 'Năm',
       'GIOI_TINH', 'SO_CCCD', 'TINH', 'HUYEN', 'XA', 'DANTOC', 'PHAN_LOAI_HO',
       'KET_QUA', 'GHI_CHU', 'DC_DINHDANH', 'DC_CMND'],
      dtype='object')

> Xoá những người chết,đã chết

In [392]:
# List of values to filter out
values_to_drop = ["cắt hộ nghèo","ĐÃ CHẾT", "chết", "CHẾT RỒI","CHẾT", "CHUYỂN ĐI","mới chết", "CHẾT RÒI", "CHUYỂN","chuyển","Chết","chuyển đi","cắt khẩu","chuyển khẩu","CẮT CHUYỂN KHẨU","ĐÃ CHUYỂN HK","KHÔNG CÓ TẠI ĐỊA PHƯƠNG","Đã chuyển khẩu","đã chết"]

# Iterate over each column and filter rows containing values_to_drop
for column in df_huyen_dak_song.columns:
    df_huyen_dak_song = df_huyen_dak_song[~df_huyen_dak_song[column].isin(values_to_drop)]

> Kết quả sau khi loại bỏ

In [393]:
df_huyen_dak_song.shape

(8987, 22)

In [394]:
df_huyen_dak_song.columns

Index(['Unnamed: 0', 'STT_HO', 'STT', 'HO_TEN_CHU_HO', 'HO_TEN_THANH_VIEN',
       'QUANHE_CHUHO', 'NGAY_SINH', 'NGAY_SINH.1', 'Ngay', 'Tháng', 'Năm',
       'GIOI_TINH', 'SO_CCCD', 'TINH', 'HUYEN', 'XA', 'DANTOC', 'PHAN_LOAI_HO',
       'KET_QUA', 'GHI_CHU', 'DC_DINHDANH', 'DC_CMND'],
      dtype='object')

> Các cột như DC_CMND, DC_DinhDanh,Ghi_Chu,KetQua,Unnamed: 0,'NGAY_SINH.1', 'Ngay', 'Tháng', 'Năm' không có ý nghĩa phân tích nên ta sẽ xoá bỏ 

In [395]:
features=['KET_QUA', 'GHI_CHU', 'DC_DINHDANH',
       'DC_CMND','Unnamed: 0','NGAY_SINH.1', 'Ngay', 'Tháng', 'Năm']
df_huyen_dak_song.drop(columns=features,inplace=True)

> Kết quả sau khi loại bỏ

In [396]:
df_huyen_dak_song.isnull().sum()

STT_HO               7913
STT                  2191
HO_TEN_CHU_HO           0
HO_TEN_THANH_VIEN       1
QUANHE_CHUHO            3
NGAY_SINH               0
GIOI_TINH             479
SO_CCCD               410
TINH                    0
HUYEN                   0
XA                      0
DANTOC                809
PHAN_LOAI_HO            0
dtype: int64

In [397]:
df_huyen_dak_song.shape

(8987, 13)

> Xóa bỏ giá trị trùng lặp

In [398]:
df_huyen_dak_song.drop_duplicates(inplace=True)

In [399]:
df_huyen_dak_song.shape

(8987, 13)

In [400]:
df_huyen_dak_song.isnull().sum()

STT_HO               7913
STT                  2191
HO_TEN_CHU_HO           0
HO_TEN_THANH_VIEN       1
QUANHE_CHUHO            3
NGAY_SINH               0
GIOI_TINH             479
SO_CCCD               410
TINH                    0
HUYEN                   0
XA                      0
DANTOC                809
PHAN_LOAI_HO            0
dtype: int64

> Điền giá trị phổ biến cho cột dân tộc theo xã

In [401]:
df_huyen_dak_song=fill_missing_dantoc(df_huyen_dak_song)

> Điền giá trị phổ biến cho cột giới tính


In [402]:
# Tính giá trị phổ biến nhất trong cột GIOI_TINH
most_common_gender = df_huyen_dak_song['GIOI_TINH'].mode()[0]

# Điền giá trị phổ biến vào các ô trống
df_huyen_dak_song.fillna({'GIOI_TINH': most_common_gender}, inplace=True)


> Tính tuổi dựa trên ngày sinh

In [403]:
df_huyen_dak_song['TUOI']=round(df_huyen_dak_song['NGAY_SINH'].apply(calculate_age))

In [404]:
df_huyen_dak_song.isnull().sum()

STT_HO               7913
STT                  2191
HO_TEN_CHU_HO           0
HO_TEN_THANH_VIEN       1
QUANHE_CHUHO            3
NGAY_SINH               0
GIOI_TINH               0
SO_CCCD               410
TINH                    0
HUYEN                   0
XA                      0
DANTOC                  0
PHAN_LOAI_HO            0
TUOI                  154
dtype: int64

> Điền giá trị trung bình cho các hàng không tính tuổi được

In [405]:
# Tính giá trị trung bình trong cột Tuổi
most_common_year_old = df_huyen_dak_song['TUOI'].mean()

# Điền giá trị trung bình  vào các ô trống
df_huyen_dak_song.fillna({'TUOI': most_common_year_old }, inplace=True)

> Điền giá trị phổ biến cho cột QUANHE_CHUHO theo từng xã

In [406]:
df_huyen_dak_song = fill_missing_quanhe_chuho(df_huyen_dak_song)

> Ban đầu ta có (9028, 22) sau quá trình tiền xử lí thì (8618, 14)

In [407]:
df_huyen_dak_song.shape

(8987, 14)

In [408]:
df_huyen_dak_song.isnull().sum()

STT_HO               7913
STT                  2191
HO_TEN_CHU_HO           0
HO_TEN_THANH_VIEN       1
QUANHE_CHUHO            0
NGAY_SINH               0
GIOI_TINH               0
SO_CCCD               410
TINH                    0
HUYEN                   0
XA                      0
DANTOC                  0
PHAN_LOAI_HO            0
TUOI                    0
dtype: int64

In [409]:
df_huyen_dak_song['XA'].unique()

array(['ĐẮK HÒA', 'TRƯỜNG XUÂN', 'THUẬN\nHẠNH', "ĐẮK N'DRUNG", 'ĐẮK MÔL',
       'THUẬN HÀ', 'NAM BÌNH', "NÂM N'JANG", 'ĐỨC AN'], dtype=object)

> Chuyển đổi các thuộc tính sang biến phân loại

In [410]:
gioi_tinh_mapping = {
    'FEMALE': 2,
    'MALE': 1
}

# Thay đổi giá trị trong cột 'GIOI_TINH' theo từ điển ánh xạ
df_huyen_dak_song['GIOI_TINH'] = df_huyen_dak_song['GIOI_TINH'].map(gioi_tinh_mapping)

In [411]:
mapping_dan_toc = {
    'KIN': 1,
    'NUN': 7,
    'TAY': 2,
    'MNO': 20,
    'MUO': 6,
    'GIR': 10,
    'THAI': 3,
    'KHM': 5,
    'DAO': 9,
    'HMO': 8,
    'THA': 3,
    'CHA': 17,
    'SAC': 15,
    'MAA': 28,
    'EDE': 12,
    'HOA': 4,
    'COH': 16,
    'THO': 24,
    'SAD': 18,
    'XTI': 22,
    'HRE': 19,
    'XED': 14,
    '56':56
}

# Thay đổi giá trị trong cột 'DANTOC'
df_huyen_dak_song['DANTOC'] = df_huyen_dak_song['DANTOC'].map(mapping_dan_toc)

In [412]:
df_huyen_dak_song.shape

(8987, 14)

In [413]:
mapping_phan_loai_ho={
    'POOR':1,
    'NEAR_POOR':2
}
df_huyen_dak_song['PHAN_LOAI_HO']=df_huyen_dak_song['PHAN_LOAI_HO'].map(mapping_phan_loai_ho)

> Xuất File

In [414]:
df_huyen_dak_song.to_excel("../data_processed/HUYỆN ĐẮK SONG.xlsx")

#### 2.1.6. HUYỆN KRÔNG NÔ

> Kiểm tra giá trị null

In [415]:
df_huyen_krong_no.isnull().sum()

b                        0
STT_HO               11797
STT                      0
HO_TEN_CHU_HO            0
HO_TEN_THANH_VIEN        5
QUANHE_CHUHO           747
NGAY_SINH               29
Ngay                  1267
tháng                 1269
năm                    279
GIOI_TINH             1081
SO_CCCD                670
TINH                     0
HUYEN                    0
XA                       0
DANTOC                 863
PHAN_LOAI_HO             0
KET_QUA                 12
GHI_CHU              13271
DC_DINHDANH           2872
DC_CMND               8075
dtype: int64

In [416]:
df_huyen_krong_no.shape

(15094, 21)

In [417]:
df_huyen_krong_no.columns

Index(['b', 'STT_HO', 'STT', 'HO_TEN_CHU_HO', 'HO_TEN_THANH_VIEN',
       'QUANHE_CHUHO', 'NGAY_SINH', 'Ngay', 'tháng', 'năm', 'GIOI_TINH',
       'SO_CCCD', 'TINH', 'HUYEN', 'XA', 'DANTOC', 'PHAN_LOAI_HO', 'KET_QUA',
       'GHI_CHU', 'DC_DINHDANH', 'DC_CMND'],
      dtype='object')

> Quan sát ta thấy những người có số CCCD bị trống sẽ không thể tìm thấy trên hệ thống nên drop bỏ những người đó ra khỏi bộ dữ liệu

In [418]:
# List of values to filter out
values_to_drop = ["cắt hộ nghèo","ĐÃ CHẾT", "chết", "CHẾT RỒI","CHẾT", "CHUYỂN ĐI","mới chết", "CHẾT RÒI", "CHUYỂN","chuyển","Chết","chuyển đi","cắt khẩu","chuyển khẩu","CẮT CHUYỂN KHẨU","ĐÃ CHUYỂN HK","KHÔNG CÓ TẠI ĐỊA PHƯƠNG","Đã chuyển khẩu","đã chết"]
# Iterate over each column and filter rows containing values_to_drop
for column in df_huyen_krong_no.columns:
    df_huyen_krong_no =df_huyen_krong_no[~df_huyen_krong_no[column].isin(values_to_drop)]

> Kết quả sau khi loại bỏ

In [419]:
df_huyen_krong_no.shape

(15073, 21)

In [420]:
df_huyen_krong_no.columns

Index(['b', 'STT_HO', 'STT', 'HO_TEN_CHU_HO', 'HO_TEN_THANH_VIEN',
       'QUANHE_CHUHO', 'NGAY_SINH', 'Ngay', 'tháng', 'năm', 'GIOI_TINH',
       'SO_CCCD', 'TINH', 'HUYEN', 'XA', 'DANTOC', 'PHAN_LOAI_HO', 'KET_QUA',
       'GHI_CHU', 'DC_DINHDANH', 'DC_CMND'],
      dtype='object')

> Các cột như DC_CMND, DC_DinhDanh,Ghi_Chu,KetQua,b,Ngay,tháng,năm  không có ý nghĩa phân tích nên ta sẽ xoá bỏ 

In [421]:
features=['KET_QUA','GHI_CHU', 'DC_DINHDANH', 'DC_CMND','b','Ngay', 'tháng', 'năm']
df_huyen_krong_no.drop(columns=features,inplace=True)

In [422]:
df_huyen_krong_no.shape

(15073, 13)

> Loại bỏ giá trị trùng lặp

In [423]:
df_huyen_krong_no.drop_duplicates(inplace=True)

In [424]:
df_huyen_krong_no.shape

(15073, 13)

In [425]:
df_huyen_krong_no.isnull().sum()

STT_HO               11778
STT                      0
HO_TEN_CHU_HO            0
HO_TEN_THANH_VIEN        5
QUANHE_CHUHO           747
NGAY_SINH               29
GIOI_TINH             1081
SO_CCCD                665
TINH                     0
HUYEN                    0
XA                       0
DANTOC                 862
PHAN_LOAI_HO             0
dtype: int64

> Điền giá trị phổ biến cho cột dân tộc theo xã

In [426]:
df_huyen_krong_no=fill_missing_dantoc(df_huyen_krong_no)

> Điền giá trị phổ biến cho cột giới tính

In [427]:
# Tính giá trị phổ biến nhất trong cột GIOI_TINH
most_common_gender = df_huyen_krong_no['GIOI_TINH'].mode()[0]

# Điền giá trị phổ biến vào các ô trống
df_huyen_krong_no.fillna({'GIOI_TINH': most_common_gender}, inplace=True)

> Tính tuổi dựa trên ngày sinh

In [428]:
df_huyen_krong_no['TUOI']=round(df_huyen_krong_no['NGAY_SINH'].apply(calculate_age))

In [429]:
df_huyen_krong_no.isnull().sum()

STT_HO               11778
STT                      0
HO_TEN_CHU_HO            0
HO_TEN_THANH_VIEN        5
QUANHE_CHUHO           747
NGAY_SINH               29
GIOI_TINH                0
SO_CCCD                665
TINH                     0
HUYEN                    0
XA                       0
DANTOC                   0
PHAN_LOAI_HO             0
TUOI                   441
dtype: int64

> Điền giá trị trung bình cho các hàng không tính tuổi được

In [430]:
# Tính giá trị trung bình trong cột Tuổi
most_common_year_old = df_huyen_krong_no['TUOI'].mean()

# Điền giá trị trung bình  vào các ô trống
df_huyen_krong_no.fillna({'TUOI': most_common_year_old }, inplace=True)

> Điền giá trị phổ biến cho cột QUANHE_CHUHO theo từng xã

In [431]:
df_huyen_krong_no = fill_missing_quanhe_chuho(df_huyen_krong_no)

> Ban đầu ta có (15094, 21) sau khi tiền xử lí thì còn lại (14424, 14)

In [432]:
df_huyen_krong_no.shape

(15073, 14)

In [433]:
df_huyen_krong_no.isnull().sum()

STT_HO               11778
STT                      0
HO_TEN_CHU_HO            0
HO_TEN_THANH_VIEN        5
QUANHE_CHUHO             0
NGAY_SINH               29
GIOI_TINH                0
SO_CCCD                665
TINH                     0
HUYEN                    0
XA                       0
DANTOC                   0
PHAN_LOAI_HO             0
TUOI                     0
dtype: int64

In [434]:
df_huyen_krong_no['XA'].unique()

array(['ĐẮK MÂM', 'QUẢNG PHÚ', 'ĐĂK NANG', 'ĐỨC XUYÊN', "NÂM N'ĐIR",
       'NÂM NUNG', 'TÂN THÀNH', 'ĐẮK DRÔ', 'BUÔN CHOAH', 'NAM ĐÀ',
       'ĐẮK SÔR', 'NAM XUÂN'], dtype=object)

In [435]:
df_huyen_krong_no[df_huyen_krong_no['PHAN_LOAI_HO'].isnull()]

Empty DataFrame
Columns: [STT_HO, STT, HO_TEN_CHU_HO, HO_TEN_THANH_VIEN, QUANHE_CHUHO, NGAY_SINH, GIOI_TINH, SO_CCCD, TINH, HUYEN, XA, DANTOC, PHAN_LOAI_HO, TUOI]
Index: []

In [436]:
df_huyen_krong_no.dropna(subset='PHAN_LOAI_HO',inplace=True)

> Chuyển đổi các thuộc tính sang biến phân loại

In [437]:
gioi_tinh_mapping = {
    'FEMALE': 2,
    'MALE': 1
}

# Thay đổi giá trị trong cột 'GIOI_TINH' theo từ điển ánh xạ
df_huyen_krong_no['GIOI_TINH'] = df_huyen_krong_no['GIOI_TINH'].map(gioi_tinh_mapping)

In [438]:
mapping_phan_loai_ho={
    'POOR':1,
    'NEAR_POOR':2
}
df_huyen_krong_no['PHAN_LOAI_HO']=df_huyen_krong_no['PHAN_LOAI_HO'].map(mapping_phan_loai_ho)

In [439]:
mapping_dan_toc = {
    'KIN': 1, 'NUN': 7, 'TAY': 2, 'MNO': 20, 'MUO': 6, 'GIR': 10, 'THA': 3,
    'KHM': 5, 'DAO': 9, 'HMO': 8, 'EDE': 12, 'SAC': 15, 'HOA': 4, 'SAD': 18,
    'THO': 24,'NGA':11
}

# Thực hiện thay đổi giá trị trong cột 'DANTOC' của DataFrame
df_huyen_krong_no['DANTOC'] = df_huyen_krong_no['DANTOC'].map(mapping_dan_toc)

> Xuất file

In [440]:
df_huyen_krong_no.to_excel("../data_processed/HUYỆN KRÔNG NÔ.xlsx")

#### 2.1.7.HUYỆN TUY ĐỨC

> Kiểm tra giá trị null

In [441]:
df_huyen_tuy_duc.isnull().sum()

Unnamed: 0               0
STT_HO               13309
STT                      0
HO_TEN_CHU_HO            0
HO_TEN_THANH_VIEN       22
QUANHE_CHUHO            17
NGAY_SINH               18
NGAY_SINH.1             18
NGAY                  1002
Tháng                 1003
Năm                     33
GIOI_TINH                1
SO_CCCD                129
TINH                     0
HUYEN                    0
XA                       0
THÔN XÓM             22178
DANTOC                  24
PHAN_LOAI_HO             0
KET_QUA                  0
GHI_CHU              20323
DC_DINHDANH           7423
DC_CMND              12667
Unnamed: 23          18008
dtype: int64

In [442]:
df_huyen_tuy_duc.shape

(22178, 24)

In [443]:
df_huyen_tuy_duc.columns

Index(['Unnamed: 0', 'STT_HO', 'STT', 'HO_TEN_CHU_HO', 'HO_TEN_THANH_VIEN',
       'QUANHE_CHUHO', 'NGAY_SINH', 'NGAY_SINH.1', 'NGAY', 'Tháng', 'Năm',
       'GIOI_TINH', 'SO_CCCD', 'TINH', 'HUYEN', 'XA', 'THÔN XÓM', 'DANTOC',
       'PHAN_LOAI_HO', 'KET_QUA', 'GHI_CHU', 'DC_DINHDANH', 'DC_CMND',
       'Unnamed: 23'],
      dtype='object')

> Quan sát ta thấy những người có số CCCD bị trống sẽ không thể tìm thấy trên hệ thống nên drop bỏ những người đó ra khỏi bộ dữ liệu

In [444]:
# List of values to filter out
values_to_drop = ["cắt hộ nghèo","ĐÃ CHẾT", "chết", "CHẾT RỒI","CHẾT", "CHUYỂN ĐI","mới chết", "CHẾT RÒI", "CHUYỂN","chuyển","Chết","chuyển đi","cắt khẩu","chuyển khẩu","CẮT CHUYỂN KHẨU","ĐÃ CHUYỂN HK","KHÔNG CÓ TẠI ĐỊA PHƯƠNG","Đã chuyển khẩu","đã chết"]
# Iterate over each column and filter rows containing values_to_drop
for column in df_huyen_tuy_duc.columns:
    df_huyen_tuy_duc = df_huyen_tuy_duc[~df_huyen_tuy_duc[column].isin(values_to_drop)]

> Kết quả sau khi xóa

In [445]:
df_huyen_tuy_duc.shape

(22169, 24)

In [446]:
df_huyen_tuy_duc.columns

Index(['Unnamed: 0', 'STT_HO', 'STT', 'HO_TEN_CHU_HO', 'HO_TEN_THANH_VIEN',
       'QUANHE_CHUHO', 'NGAY_SINH', 'NGAY_SINH.1', 'NGAY', 'Tháng', 'Năm',
       'GIOI_TINH', 'SO_CCCD', 'TINH', 'HUYEN', 'XA', 'THÔN XÓM', 'DANTOC',
       'PHAN_LOAI_HO', 'KET_QUA', 'GHI_CHU', 'DC_DINHDANH', 'DC_CMND',
       'Unnamed: 23'],
      dtype='object')

> Các cột như 'KET_QUA', 'GHI_CHU', 'DC_DINHDANH', 'DC_CMND',
       'Unnamed: 23','Unnamed: 0','NGAY_SINH.1','NGAY', 'Tháng', 'Năm','THÔN XÓM' không có ý nghĩa phân tích nên ta sẽ xoá bỏ 

In [447]:
features=['KET_QUA', 'GHI_CHU', 'DC_DINHDANH', 'DC_CMND',
       'Unnamed: 23','Unnamed: 0','NGAY_SINH.1','NGAY', 'Tháng', 'Năm','THÔN XÓM']
df_huyen_tuy_duc.drop(columns=features,inplace=True)

In [448]:
df_huyen_tuy_duc.shape

(22169, 13)

> Xóa bỏ giá trị trùng lặp

In [449]:
df_huyen_tuy_duc.drop_duplicates(inplace=True)

In [450]:
df_huyen_tuy_duc.shape

(22168, 13)

In [451]:
df_huyen_tuy_duc.isnull().sum()

STT_HO               13307
STT                      0
HO_TEN_CHU_HO            0
HO_TEN_THANH_VIEN       22
QUANHE_CHUHO            17
NGAY_SINH               18
GIOI_TINH                1
SO_CCCD                128
TINH                     0
HUYEN                    0
XA                       0
DANTOC                  24
PHAN_LOAI_HO             0
dtype: int64

> Điền giá trị phổ biến cho cột dân tộc theo xã

In [452]:
df_huyen_tuy_duc=fill_missing_dantoc(df_huyen_tuy_duc)

> Điền giá trị phổ biến cho cột giới tính

In [453]:
# Tính giá trị phổ biến nhất trong cột GIOI_TINH
most_common_gender = df_huyen_tuy_duc['GIOI_TINH'].mode()[0]

# Điền giá trị phổ biến vào các ô trống
df_huyen_tuy_duc.fillna({'GIOI_TINH': most_common_gender}, inplace=True)

> Tính tuổi dựa trên ngày sinh

In [454]:
df_huyen_tuy_duc['TUOI']=round(df_huyen_tuy_duc['NGAY_SINH'].apply(calculate_age))


In [455]:
df_huyen_tuy_duc.isnull().sum()

STT_HO               13307
STT                      0
HO_TEN_CHU_HO            0
HO_TEN_THANH_VIEN       22
QUANHE_CHUHO            17
NGAY_SINH               18
GIOI_TINH                0
SO_CCCD                128
TINH                     0
HUYEN                    0
XA                       0
DANTOC                   0
PHAN_LOAI_HO             0
TUOI                   377
dtype: int64

> Điền giá trị trung bình cho các hàng không tính tuổi được

In [456]:
# Tính giá trị trung bình trong cột Tuổi
most_common_year_old = df_huyen_tuy_duc['TUOI'].mean()

# Điền giá trị trung bình  vào các ô trống
df_huyen_tuy_duc.fillna({'TUOI': most_common_year_old }, inplace=True)

> Điền giá trị phổ biến cho cột QUANHE_CHUHO theo từng xã

In [457]:
df_huyen_tuy_duc = fill_missing_quanhe_chuho(df_huyen_tuy_duc)

> Ban đầu ta có (22181, 24) sau khi tiền xử lí dữ liệu thì còn (22048, 14)

In [458]:
df_huyen_tuy_duc.shape

(22168, 14)

In [459]:
df_huyen_tuy_duc[df_huyen_tuy_duc['PHAN_LOAI_HO'].isnull()]

Empty DataFrame
Columns: [STT_HO, STT, HO_TEN_CHU_HO, HO_TEN_THANH_VIEN, QUANHE_CHUHO, NGAY_SINH, GIOI_TINH, SO_CCCD, TINH, HUYEN, XA, DANTOC, PHAN_LOAI_HO, TUOI]
Index: []

In [460]:
df_huyen_tuy_duc.isnull().sum()

STT_HO               13307
STT                      0
HO_TEN_CHU_HO            0
HO_TEN_THANH_VIEN       22
QUANHE_CHUHO             0
NGAY_SINH               18
GIOI_TINH                0
SO_CCCD                128
TINH                     0
HUYEN                    0
XA                       0
DANTOC                   0
PHAN_LOAI_HO             0
TUOI                     0
dtype: int64

> Chuyển đổi các thuộc tính sang biến phân loại

In [461]:
mapping_phan_loai_ho={
    'POOR':1,
    'NEAR_POOR':2
}
df_huyen_tuy_duc['PHAN_LOAI_HO']=df_huyen_tuy_duc['PHAN_LOAI_HO'].map(mapping_phan_loai_ho)

In [462]:
gioi_tinh_mapping = {
    'FEMALE': 2,
    'MALE': 1
}

# Thay đổi giá trị trong cột 'GIOI_TINH' theo từ điển ánh xạ
df_huyen_tuy_duc['GIOI_TINH'] = df_huyen_tuy_duc['GIOI_TINH'].map(gioi_tinh_mapping)

In [463]:
df_huyen_tuy_duc['XA'].unique()

array(['QUẢNG TÂN', 'QUẢNG TRỰC', 'QUẢNG TÂM', 'ĐẮK BÚK SO', 'ĐẮK NGO',
       "ĐẮ R'TÍH"], dtype=object)

In [464]:
# Định nghĩa từ điển ánh xạ giá trị cũ sang giá trị mới
mapping_dan_toc = {
    'KIN': 1, 'NUN': 7, 'TAY': 2, 'MNO': 20, 'MUO': 6, 'GIR': 10, 'THA': 3,
    'KHM': 5, 'DAO': 9, 'HMO': 8, 'EDE': 12, 'SAC': 15, 'HOA': 4, 'SAD': 18,
    'THO': 24, 'COO': 30, 'CHO': 32, 'RAG': 21, 'XTI': 22, 'MAA': 28, 'CHA': 17,
    'LAO': 37, 'BAN': 13, 'CHR': 36, 'XED': 14,'56':56
}

# Thực hiện thay đổi giá trị trong cột 'DANTOC' của DataFrame
df_huyen_tuy_duc['DANTOC'] = df_huyen_tuy_duc['DANTOC'].map(mapping_dan_toc)


> Xuất file

In [465]:
df_huyen_tuy_duc.to_excel("../data_processed/HUYỆN TUY ĐỨC.xlsx")

#### 2.1.8. TP GIA NGHĨA.xlsx

> Kiểm tra giá trị null

In [466]:
df_tp_gia_nghia.isnull().sum()

Unnamed: 0             0
STT_HO               728
STT                    0
HO_TEN_CHU_HO          0
HO_TEN_THANH_VIEN      0
QUANHE_CHUHO           0
NGAY_SINH              4
NGAY_SINH.1            3
Ngày                  15
Tháng                 15
Năm                    3
GIOI_TINH              2
SO_CCCD               66
TINH                   0
HUYEN                  0
XA                     0
Thôn, Bon, TDP         0
DANTOC                 0
PHAN_LOAI_HO           0
KET_QUA                0
GHI_CHU              727
DC_DINHDANH          150
DC_CMND              501
Unnamed: 23          947
Unnamed: 24          899
dtype: int64

In [467]:
df_tp_gia_nghia.shape

(969, 25)

In [468]:
df_tp_gia_nghia.columns

Index(['Unnamed: 0', 'STT_HO', 'STT', 'HO_TEN_CHU_HO', 'HO_TEN_THANH_VIEN',
       'QUANHE_CHUHO', 'NGAY_SINH', 'NGAY_SINH.1', 'Ngày', 'Tháng', 'Năm',
       'GIOI_TINH', 'SO_CCCD', 'TINH', 'HUYEN', 'XA', 'Thôn, Bon, TDP',
       'DANTOC', 'PHAN_LOAI_HO', 'KET_QUA', 'GHI_CHU', 'DC_DINHDANH',
       'DC_CMND', 'Unnamed: 23', 'Unnamed: 24'],
      dtype='object')

> Xoá những người đã chết,cắt hộ nghèo, chuyển vùng

In [469]:
# List of values to filter out
values_to_drop = ["cắt hộ nghèo","ĐÃ CHẾT", "chết", "CHẾT RỒI","CHẾT", "CHUYỂN ĐI","mới chết", "CHẾT RÒI", "CHUYỂN","chuyển","Chết","chuyển đi","cắt khẩu","chuyển khẩu","CẮT CHUYỂN KHẨU","ĐÃ CHUYỂN HK","KHÔNG CÓ TẠI ĐỊA PHƯƠNG","Đã chuyển khẩu","đã chết"]
# Iterate over each column and filter rows containing values_to_drop
for column in df_tp_gia_nghia.columns:
    df_tp_gia_nghia = df_tp_gia_nghia[~df_tp_gia_nghia[column].isin(values_to_drop)]

> Kết quả sau khi xóa bỏ

In [470]:
df_tp_gia_nghia.shape

(969, 25)

In [471]:
df_tp_gia_nghia.columns

Index(['Unnamed: 0', 'STT_HO', 'STT', 'HO_TEN_CHU_HO', 'HO_TEN_THANH_VIEN',
       'QUANHE_CHUHO', 'NGAY_SINH', 'NGAY_SINH.1', 'Ngày', 'Tháng', 'Năm',
       'GIOI_TINH', 'SO_CCCD', 'TINH', 'HUYEN', 'XA', 'Thôn, Bon, TDP',
       'DANTOC', 'PHAN_LOAI_HO', 'KET_QUA', 'GHI_CHU', 'DC_DINHDANH',
       'DC_CMND', 'Unnamed: 23', 'Unnamed: 24'],
      dtype='object')

> Các cột như Thôn, Bon, TDP,'KET_QUA', 'GHI_CHU', 'DC_DINHDANH',
       'DC_CMND', 'Unnamed: 23', 'Unnamed: 24','Unnamed: 0','NGAY_SINH.1', 'Ngày', 'Tháng', 'Năm' không có ý nghĩa phân tích nên ta sẽ xoá bỏ 

In [472]:
features=['Thôn, Bon, TDP','KET_QUA', 'GHI_CHU', 'DC_DINHDANH',
       'DC_CMND', 'Unnamed: 23', 'Unnamed: 24','Unnamed: 0','NGAY_SINH.1', 'Ngày', 'Tháng', 'Năm']
df_tp_gia_nghia.drop(columns=features,inplace=True)

In [473]:
df_tp_gia_nghia.shape

(969, 13)

> Xóa giá trị trùng lặp

In [474]:
df_tp_gia_nghia.drop_duplicates(inplace=True)

In [475]:
df_tp_gia_nghia.shape

(969, 13)

In [476]:
df_tp_gia_nghia.isnull().sum()

STT_HO               728
STT                    0
HO_TEN_CHU_HO          0
HO_TEN_THANH_VIEN      0
QUANHE_CHUHO           0
NGAY_SINH              4
GIOI_TINH              2
SO_CCCD               66
TINH                   0
HUYEN                  0
XA                     0
DANTOC                 0
PHAN_LOAI_HO           0
dtype: int64

> Điền giá trị phổ biến cho cột giới tính

In [477]:
# Tính giá trị phổ biến nhất trong cột GIOI_TINH
most_common_gender = df_tp_gia_nghia['GIOI_TINH'].mode()[0]

# Điền giá trị phổ biến vào các ô trống
df_tp_gia_nghia.fillna({'GIOI_TINH': most_common_gender}, inplace=True)

In [478]:
df_tp_gia_nghia.shape

(969, 13)

> Tính tuổi dựa trên ngày sinh

In [479]:
# Chuyển cột NGAY_SINH thành kiểu datetime với xử lý lỗi 'coerce'
df_tp_gia_nghia['NGAY_SINH'] = pd.to_datetime(df_tp_gia_nghia['NGAY_SINH'], errors='coerce')

# Loại bỏ các dòng có NGAY_SINH là NaT
df_tp_gia_nghia = df_tp_gia_nghia.dropna(subset=['NGAY_SINH'])

# Chuyển đổi NGAY_SINH thành chuỗi yyyy-mm-dd
df_tp_gia_nghia['NGAY_SINH'] = df_tp_gia_nghia['NGAY_SINH'].dt.strftime('%Y-%m-%d')

# Tính tuổi từ ngày sinh
def calculate_age_tp(birthdate):
    today = datetime.today()
    age = today.year - birthdate.year - ((today.month, today.day) < (birthdate.month, birthdate.day))
    return age

# Áp dụng hàm tính tuổi vào dataframe
df_tp_gia_nghia['TUOI'] = round(df_tp_gia_nghia['NGAY_SINH'].apply(lambda x: calculate_age_tp(datetime.strptime(x, '%Y-%m-%d'))))

In [480]:
df_tp_gia_nghia.shape

(958, 14)

In [481]:
df_tp_gia_nghia.isnull().sum()

STT_HO               721
STT                    0
HO_TEN_CHU_HO          0
HO_TEN_THANH_VIEN      0
QUANHE_CHUHO           0
NGAY_SINH              0
GIOI_TINH              0
SO_CCCD               66
TINH                   0
HUYEN                  0
XA                     0
DANTOC                 0
PHAN_LOAI_HO           0
TUOI                   0
dtype: int64

> Điền giá trị phổ biến cho cột QUANHE_CHUHO theo từng xã

In [482]:
df_tp_gia_nghia = fill_missing_quanhe_chuho(df_tp_gia_nghia )

> Chuyển đổi các thuộc tính sang biến phân loại

In [483]:
gioi_tinh_mapping = {
    'FEMALE': 2,
    'MALE': 1
}

# Thay đổi giá trị trong cột 'GIOI_TINH' theo từ điển ánh xạ
df_tp_gia_nghia['GIOI_TINH'] = df_tp_gia_nghia['GIOI_TINH'].map(gioi_tinh_mapping)

In [484]:
mapping_phan_loai_ho={
    'POOR':1,
    'NEAR_POOR':2
}
df_tp_gia_nghia['PHAN_LOAI_HO']=df_tp_gia_nghia['PHAN_LOAI_HO'].map(mapping_phan_loai_ho)

In [485]:
mapping_dan_toc = {
    'KIN': 1,
    'TAY': 2,
    'THA': 3,
    'MNO': 20,
    'HOA': 4,
    'NUN': 7,
    'THO': 24,
    'MAA': 28,
    'EDE': 12,
    'KHM': 5,
    'DAO': 9,
    'HMO': 8,
    'SAC': 15,
    'CHU': 44,
    'MUO': 6
}

df_tp_gia_nghia['DANTOC'] = df_tp_gia_nghia['DANTOC'].map(mapping_dan_toc)

> Ban đầu ta có (969, 25) sau khi tiền xử lí thì còn lại (892, 14)

In [486]:
df_tp_gia_nghia.shape

(958, 14)

In [487]:
df_tp_gia_nghia.isnull().sum()

STT_HO               721
STT                    0
HO_TEN_CHU_HO          0
HO_TEN_THANH_VIEN      0
QUANHE_CHUHO           0
NGAY_SINH              0
GIOI_TINH              0
SO_CCCD               66
TINH                   0
HUYEN                  0
XA                     0
DANTOC                 0
PHAN_LOAI_HO           0
TUOI                   0
dtype: int64

> Xuất File

In [488]:
df_tp_gia_nghia.to_excel("../data_processed/TP GIA NGHĨA.xlsx")

>

> 